usage = "
   Summarizes the number of non-structural reads for each the libraries. It uses the output from htseq-count.
   Inputs
   1: table with gene names and corresponding biotypes
   2: '.counts' files"

libs=c(
  "data.table",
  # "RColorBrewer",
  # "ggplot2",
  "DESeq2",
  # "scales",
  # "pheatmap",
  "org.Ce.eg.db",
  "stringr"
)
lapply(libs, require, character.only=T)


args <- commandArgs(trailingOnly = TRUE)
# tab="featureCounts_summary.txt"
# bio_file="/home/adomingu/imb-kettinggr/genomes/Caenorhabditis_elegans/Ensembl/WBcel235/Annotation/Genes/wormbaseGeneID2biotype.txt"
# tab <- args[1]
bio_file <- args[1]
count_files <- args[2:length(args)]


# sample_files <- grep(".counts", list.files(dir), value=TRUE)
prepareCountMatrix <- function(sample_files=count_files){
  # reads files generated by htseq-counts
  # creates a tmp DESeqDataSet
  # returns the count matrix
  
  sample_name <- gsub('.counts', "", basename(sample_files))

  sample_name  <- ifelse(
    str_detect(sample_name, '\\.\\d+'),
    sample_name,
    paste0(sample_name, ".All")
  )

  conds <- gsub(
     "(^\\w+)_\\w+_r\\d+_\\w+.\\w+.counts", "\\1",
     sample_files
     )

  sample_table <- data.frame(
    sampleName = sample_name,
    fileName = basename(sample_files),
    condition = sample_name)

  dds <- DESeqDataSetFromHTSeqCount(
    sampleTable = sample_table,
    directory = dirname(sample_files[1]),
    design= ~ condition)

  countdata_m <- counts(dds)
  return(countdata_m)
}

countdata <- setDT(
   data.frame(
      prepareCountMatrix()),
      keep.rownames = TRUE)[]
setnames(countdata, "rn", "wormbase_gene")


# gene_features <- read.table(tab, sep="\t", header=T)
# nc <- ncol(gene_features)
# colnames(gene_features)[1] <- "wormbase_gene"
# ## remove cols with strand, start...
# countdata <- gene_features[,c(1,7:nc)]

## clean RNA classes
colnames(countdata) <- gsub(
   ".+\\.(\\w+.2\\d\\w).bam",
   "\\1",
   colnames(countdata))

colnames(countdata) <- gsub(
   ".+\\.(\\w+).bam",
   "\\1",
   colnames(countdata)
   )

countdata <- as.data.frame(countdata)
## select full libs
countdata <- countdata[!grepl('\\w+.2\\d\\w$', colnames(countdata))]
countdata$wormbase_gene <- rownames(countdata)
# head(countdata)

## biotype:
bio <- read.table(bio_file,
   sep="\t")

colnames(bio) <- c("wormbase_gene", "gene_biotype")
# unique(bio$gene_biotype)

idx <- match( countdata$wormbase_gene, bio$wormbase_gene )
countdata$gene_biotype <- bio$gene_biotype[ idx ]

struct_rna <- c("rRNA", "tRNA", "snoRNA", "snRNA")
non_struct <- subset(countdata, (!gene_biotype %in% struct_rna))[, -c(1,ncol(countdata))]
# head(non_struct)

fac <- colSums(non_struct[ ,c(1:ncol(non_struct))])
# fac2 <- colSums(countdata[ ,c(2:ncol(non_struct))])

write.table(t(data.frame(as.list(fac))),
   "normalization_factors.txt",
   quote=FALSE,
   sep="\t",
   row.names=TRUE,
   col.names=FALSE
   )
