# Differential gene expression analysis


```{r packages}
######################
## packages
######################
libs=c(
  "data.table",
  "RColorBrewer",
  "ggplot2",
  "DESeq2",
  "scales",
  "pheatmap",
  "org.Ce.eg.db"
)

lapply(libs, require, character.only=T)

theme_set(theme_bw(16))
theme_white <- function() {
     theme_update(
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank()
      )
 }
theme_white()
# make a palette the default: http://stackoverflow.com/a/16437625/1274242
ggplot <- function(...) ggplot2::ggplot(...) + scale_color_brewer(palette="Set1")
dir.create(file.path("figure"), showWarnings = FALSE)

```

```{r functions}
concat <- function(...) paste(..., sep="")

df2matrix <- function(df){
# converts a dataframe in matrix
# using the 1st column as rownnames
  # # test
  # df=countdata
  mat <- as.matrix(df[, -1])
  rownames(mat) <- df[, 1]
  class(mat) <- "numeric"
  return(mat)
}

```

## Transposons: DESeq2 analysis

Transposon reads were counted separately using the annotations from wormbase.

### All small RNA types

```{r load_counts_Tranposons}
gene_features <- read.table(
    "featureCounts_summary.txt",
   sep="\t",
   header=T,
   stringsAsFactors=FALSE
)
nc <- ncol(gene_features)
colnames(gene_features)[1] <- "Geneid"
## remove cols with strand, start
countdata <- gene_features[,c(1,7:nc)]

## clean RNA classes
colnames(countdata) <- gsub(
   ".+\\.(\\w+.2\\d\\w).bam",
   "\\1",
   colnames(countdata))

## cleans path and adds "full" tag
colnames(countdata) <- gsub(
   ".+\\.(\\w+).bam",
   "\\1.All",
   colnames(countdata)
   )

```

```{r select_conditions_Tranposons}
smRNAclass <- "All"
include_bio <- "Transposons"
col_keep <- colnames(countdata)[c(grep(smRNAclass, colnames(countdata)))]
```

We tested `r nrow(countdata)` Transposons.


```{r prepare_deseq_object_Tranposons}
## filter RNA biotypes and conditions
countdata_S <- subset(
    countdata,
    select=c("Geneid", col_keep))

######################
## Convert to matrix
######################
countdata_m <- df2matrix(countdata_S)

samples <- colnames(countdata_m)

conds <- gsub(
   "(^\\w+)_\\w+_r\\d+_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
conds <- factor(
   gsub(
      "(xf)\\d+", "\\1",
      conds
      )
   )
conds

replicates <- factor(
  gsub(
   "^\\w+_\\w+_(r\\d+)_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
)

coldata <- data.frame(
  condition=conds,
  replicate=replicates,
  row.names=samples
)
coldata

## create deseq object will all the data
# in more fine-tuned analysis factors can be subseted

dds <- DESeqDataSetFromMatrix(
  countData = countdata_m,
  colData = coldata,
  design = ~ condition
)
# making sure that N2 is the first level, and thus
# log2FC are reported as treatment over control (xf/N2)
dds$condition <- relevel(dds$condition, "N2")
```

```{r deseq_analysis_Tranposons}
## DESeq
dds <- DESeq(dds)

## conditions, echo=FALSE, results="asis
col_data <- as.data.frame(colData(dds))
print(col_data)
lib_sizes <- colData(dds)$sizeFactor
```

### QC

#### Dispersion
```{r dispersion_plot_Transposons}
plotDispEsts(dds)
```


### Results

```{r results_extraction_Tranposons}
#########################################################
## Results
#########################################################
res <- results(dds)
res <- res[order(res$padj),]
summary(res)
up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

sum(res$padj < 0.1, na.rm=TRUE)
```

Summary contains an overview of the results. There are `r up` significantly up-regulated genes and `r down` significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then `r up_log2` are up-regulated and `r down_log2` are down-regulated.


```{r ma_plot_normalized_all_Transposons}
# variable position of labels
x_pos <- max(res@listData$baseMean) * 0.7
plotMA(res, main="DESeq2", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```

```{r ma_plot_not_normalized_all_Transposons}
resMLE <- results(dds, addMLE=TRUE)
# variable position of labels
x_pos <- max(resMLE@listData$baseMean) * 0.7
plotMA(resMLE, MLE=TRUE, main="unshrunken LFC", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```


```{r save_results_All_Tranposons}
fout <- paste(
   levels(conds)[2],
   "_",
   levels(conds)[1],
   "_",
   smRNAclass,
   "_",
   paste(include_bio, sep="_", collapse="_"),
   "_",
   "DESeq2_results.txt",
   sep="")
res_out <- as.data.frame(res)
res_out <- data.frame(
   wormbase_gene = rownames(res_out),
   res_out[ , !(names(res_out) == "wormbase_gene")])
# res_out$wormbase_gene <-NULL

write.table(
   res_out,
   file=fout,
   sep="\t",
   quote=FALSE,
   row.names=FALSE,
   col.names=TRUE)
```


## Transposons: DESeq2 analysis
### 22G

```{r select_conditions_22G_Tranposons}
smRNAclass <- "22G"
include_bio <- "Transposons"
col_keep <- colnames(countdata)[c(grep(smRNAclass, colnames(countdata)))]
```

We tested `r nrow(countdata)` Transposons.


```{r prepare_deseq_object_22G_Tranposons}
## filter RNA biotypes and conditions
countdata_S <- subset(
    countdata,
    select=c("Geneid", col_keep))

######################
## Convert to matrix
######################
countdata_m <- df2matrix(countdata_S)

samples <- colnames(countdata_m)

conds <- gsub(
   "(^\\w+)_\\w+_r\\d+_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
conds <- factor(
   gsub(
      "(xf)\\d+", "\\1",
      conds
      )
   )
conds

replicates <- factor(
  gsub(
   "^\\w+_\\w+_(r\\d+)_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
)

coldata <- data.frame(
  condition=conds,
  replicate=replicates,
  row.names=samples
)
coldata

## create deseq object will all the data
# in more fine-tuned analysis factors can be subseted

dds <- DESeqDataSetFromMatrix(
  countData = countdata_m,
  colData = coldata,
  design = ~ condition
)
# making sure that N2 is the first level, and thus
# log2FC are reported as treatment over control (xf/N2)
dds$condition <- relevel(dds$condition, "N2")
```

```{r deseq_analysis_22G_Tranposons}
## DESeq
dds <- DESeq(dds)

## conditions, echo=FALSE, results="asis
col_data <- as.data.frame(colData(dds))
print(col_data)
```

### QC

#### Dispersion
```{r dispersion_plot_22G_Transposons}
plotDispEsts(dds)
```


### Results

```{r results_extraction_22G_Tranposons}
#########################################################
## Results
#########################################################
res <- results(dds)
res <- res[order(res$padj),]
summary(res)
up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

sum(res$padj < 0.1, na.rm=TRUE)
```

Summary contains an overview of the results. There are `r up` significantly up-regulated genes and `r down` significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then `r up_log2` are up-regulated and `r down_log2` are down-regulated.


```{r ma_plot_normalized_22G_Transposons}
# variable position of labels
x_pos <- max(res@listData$baseMean) * 0.7
plotMA(res, main="DESeq2", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```

```{r ma_plot_not_normalized_22G_Transposons}
resMLE <- results(dds, addMLE=TRUE)
# variable position of labels
x_pos <- max(resMLE@listData$baseMean) * 0.7
plotMA(resMLE, MLE=TRUE, main="unshrunken LFC", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```


```{r save_results_22G_Tranposons}
fout <- paste(
   levels(conds)[2],
   "_",
   levels(conds)[1],
   "_",
   smRNAclass,
   "_",
   paste(include_bio, sep="_", collapse="_"),
   "_",
   "DESeq2_results.txt",
   sep="")
res_out <- as.data.frame(res)
res_out <- data.frame(
   wormbase_gene = rownames(res_out),
   res_out[ , !(names(res_out) == "wormbase_gene")])
# res_out$wormbase_gene <-NULL

write.table(
   res_out,
   file=fout,
   sep="\t",
   quote=FALSE,
   row.names=FALSE,
   col.names=TRUE)
```

## Transposons: DESeq2 analysis
### 21U


```{r select_conditions_21U_Tranposons}
smRNAclass <- "21U"
include_bio <- "Transposons"
col_keep <- colnames(countdata)[c(grep(smRNAclass, colnames(countdata)))]
```

We tested `r nrow(countdata)` Transposons.


```{r prepare_deseq_object_21U_Tranposons}
## filter RNA biotypes and conditions
countdata_S <- subset(
    countdata,
    select=c("Geneid", col_keep))

######################
## Convert to matrix
######################
countdata_m <- df2matrix(countdata_S)

samples <- colnames(countdata_m)

conds <- gsub(
   "(^\\w+)_\\w+_r\\d+_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
conds <- factor(
   gsub(
      "(xf)\\d+", "\\1",
      conds
      )
   )
conds

replicates <- factor(
  gsub(
   "^\\w+_\\w+_(r\\d+)_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
)

coldata <- data.frame(
  condition=conds,
  replicate=replicates,
  row.names=samples
)
coldata

## create deseq object will all the data
# in more fine-tuned analysis factors can be subseted

dds <- DESeqDataSetFromMatrix(
  countData = countdata_m,
  colData = coldata,
  design = ~ condition
)
# making sure that N2 is the first level, and thus
# log2FC are reported as treatment over control (xf/N2)
dds$condition <- relevel(dds$condition, "N2")
```

```{r deseq_analysis_21U_Tranposons}
## DESeq
dds <- DESeq(dds)

## conditions, echo=FALSE, results="asis
col_data <- as.data.frame(colData(dds))
print(col_data)
```

### QC

#### Dispersion
```{r dispersion_plot_21U_Transposons}
plotDispEsts(dds)
```


### Results

```{r results_extraction_21U_Tranposons}
#########################################################
## Results
#########################################################
res <- results(dds)
res <- res[order(res$padj),]
summary(res)
up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

sum(res$padj < 0.1, na.rm=TRUE)
```

Summary contains an overview of the results. There are `r up` significantly up-regulated genes and `r down` significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then `r up_log2` are up-regulated and `r down_log2` are down-regulated.


```{r ma_plot_normalized_21U_Transposons}
# variable position of labels
x_pos <- max(res@listData$baseMean) * 0.7
plotMA(res, main="DESeq2", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```

```{r ma_plot_not_normalized_21U_Transposons}
resMLE <- results(dds, addMLE=TRUE)
# variable position of labels
x_pos <- max(resMLE@listData$baseMean) * 0.7
plotMA(resMLE, MLE=TRUE, main="unshrunken LFC", ylim=c(-5,5))
text(x=x_pos, y=2, labels=up, col="blue", cex=1.5)
text(x=x_pos, y=-2, labels=down, col="blue", cex=1.5)
```


```{r save_results_21U_Tranposons}

fout <- paste(
   levels(conds)[2],
   "_",
   levels(conds)[1],
   "_",
   smRNAclass,
   "_",
   paste(include_bio, sep="_", collapse="_"),
   "_",
   "DESeq2_results.txt",
   sep="")
res_out <- as.data.frame(res)
res_out <- data.frame(
   wormbase_gene = rownames(res_out),
   res_out[ , !(names(res_out) == "wormbase_gene")])
# res_out$wormbase_gene <-NULL

write.table(
   res_out,
   file=fout,
   sep="\t",
   quote=FALSE,
   row.names=FALSE,
   col.names=TRUE)
```


## Transposons: DESeq2 analysis
### 26G


```{r select_conditions_26G_Tranposons}
smRNAclass <- "26G"
include_bio <- "Transposons"
col_keep <- colnames(countdata)[c(grep(smRNAclass, colnames(countdata)))]
```

We tested `r nrow(countdata)` Transposons.


```{r prepare_deseq_object_26G_Tranposons}
## filter RNA biotypes and conditions
countdata_S <- subset(
    countdata,
    select=c("Geneid", col_keep))

######################
## Convert to matrix
######################
countdata_m <- df2matrix(countdata_S)

samples <- colnames(countdata_m)

conds <- gsub(
   "(^\\w+)_\\w+_r\\d+_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
conds <- factor(
   gsub(
      "(xf)\\d+", "\\1",
      conds
      )
   )
conds

replicates <- factor(
  gsub(
   "^\\w+_\\w+_(r\\d+)_\\w+.\\w+", "\\1",
   colnames(countdata_m)
   )
)

coldata <- data.frame(
  condition=conds,
  replicate=replicates,
  row.names=samples
)
coldata

## create deseq object will all the data
# in more fine-tuned analysis factors can be subseted

dds <- DESeqDataSetFromMatrix(
  countData = countdata_m,
  colData = coldata,
  design = ~ condition
)
# making sure that N2 is the first level, and thus
# log2FC are reported as treatment over control (xf/N2)
dds$condition <- relevel(dds$condition, "N2")
```

```{r deseq_analysis_26G_Tranposons}
## DESeq
dds <- DESeq(dds)

## conditions, echo=FALSE, results="asis
col_data <- as.data.frame(colData(dds))
print(col_data)
```
Basically there no counts for 26G reads mapping to transposons.

```{r session_Info}
sessionInfo()
```
