---
title: "Differential gene expression of small RNA-seq in C. elegans"
author: "Ant√≥nio Domingues"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: readable
    highlight: tango
---


```{r setchunk, cache=FALSE, include=FALSE}
knitr::opts_chunk$set(
  include=TRUE,
  echo = FALSE,
  tidy=TRUE,
  fig.path = "reports/figure/",
  dev=c("png","pdf"),
  fig.align="center")
```


```{r setup, results="hide", include=FALSE}
source("DEG_analysis_DESeq2.helpers.R")

loadLibraries()

lapply(
  c("reports", "reports/figure"),
  function(x) dir.create(file.path(x), showWarnings = FALSE)
)

```


```{r input data}
tab <- "featureCounts_summary.txt"

annotation <- "/home/adomingu/imb-kettinggr/genomes/Caenorhabditis_elegans/Ensembl/WBcel235/Annotation/Genes/wormbaseGeneID2biotype.txt"

target_file <- '~/imb-kettinggr/adomingues/projects/imb_ketting_2014_14_almeida_smallRNA_celegans/data/annotations_targets/22G/all_targets.txt'

orgDB <- org.Ce.eg.db

directory <- "./"
```


# Read count

```{r prepare_counts, results="hide", include=FALSE}
countdata <- prepareCountMatrix()
dim(countdata)
```

```{r select_biotypes, results="hide", include=FALSE}
#exclude mi_pi_r_t_sn_sno_RNAs
## biotype:
countdata_m <- exludeNonStructuralReads(countdata, annotation)
dim(countdata_m)

## remove genes with very counts for performance purposes
## independent filtering is performed automatically later
countdata_m <- countdata_m[ rowSums(countdata_m) > 1, ]
dim(countdata_m)
```

Reads were assigned to genomic regions using ht-seqcount. After excluding genes with zero counts in all conditions, and mi_pi_r_t_sn_sno_RNAs, there are `r nrow(countdata_m)` genes for which the analysis will be performed.


# Analysis with all samples

This serves a QC to make sure that, for instance, library treatments cluster separately. If not, something funny happened.


```{r select_conditions, results="hide", include=FALSE}
smRNAclass <- "All"
base_name <- "all_reads_and_treatments"
countdata_sub <- selectSamples(countdata_m, c(smRNAclass))
dim(countdata_sub)
dds <- runDESeq(countdata_sub, "N2")
knitr::kable(colData(dds))
```


## QC

### Dispersion

```{r dispersion_plot}
performQC(dds, c("condition","replicate", "treatment"), base_name)
```

### Sample clustering

Here we extract the counts and transformed values to use for clustering and visualization:
- The function rlog, stands for regularized log, transforming the original count data to the log2 scale by fitting a model with a term for each sample and a prior distribution on the coefficients which is estimated from the data.
-  Variance stabilizing transformation (vsd) is a parametric fit for the dispersion.


Heatmaps showing the expression data of the 20 most highly expressed genes. The data is of log2 normalized counts, from regularized log transformation  and from variance stabilizing transformation.

Now we can actually cluster the samples:


# Results

There are many classes of _C. elegans_ small RNAs which can be enriched with different treatments. Small RNAs are split into classes according to size and nucleotide bias. We will analyze these separately.



## Untreated samples

```{r DifferentialExpression_untreated, echo=F, results='asis', error=F, warning=F, message=F}
for (treatment in c("untreated", "oxid", "TAP")){
  # subsection
  cat("\n", fill=T)
  cat("### QC", treatment, fill=T)  
  cat("\n", fill=T)
  # Prepare DESeq2 object
  smRNAclass <- "All"
  # treatment <- "untreated"
  countdata_sub <- selectSamples(countdata_m, c(smRNAclass, treatment))
  # dim(countdata_sub); summary(countdata_sub)

  dds <- runDESeq(countdata_sub, "N2")
  col_data <- as.data.frame(colData(dds))
  if (treatment == 'untreated'){
    # keep library sizes for the 26G analysis
    lib_sizes <- colData(dds)$sizeFactor
  }

  cat("A brief summary of the samples analysed:\n")
  print(knitr::kable(col_data))

  res <- results(dds)

  # ==========================================================================
  # QC
  comparison <- gsub(" ", "_",
    gsub(".*: condition ", "", res@elementMetadata$description[2])
    )

  base_name <- paste(
    smRNAclass, treatment,
    comparison,
    sep="_")
  performQC(
    dds=dds,
    groups=c("condition","replicate", "treatment"),
    base_name=base_name
    )
  cat("\n", fill=T)
  cat("These plots show the relationships between samples.
    Useful to detect outliers.")
  cat("\n", fill=T)
  cat("Assuming the QC went well, we can can not look at the results.
    ")
  cat("\n", fill=T)

  #########################################################
  ## Results
  #########################################################
  cat("\n", fill=T)
  cat("### Results", treatment, fill=T)  
  cat("\n", fill=T)

  res <- results(dds)
  res <- res[order(res$padj),]
  capture.output(summary(res))
  up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
  up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
  down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
  down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

  sum(res$padj < 0.1, na.rm=TRUE)


  cat("Summary contains an overview of the results. There are", up, " significantly up-regulated genes and", down, "significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then",  up_log2, "are up-regulated and", down_log2, "are down-regulated.
  \n")
  cat("\n", fill=T)
  
  cat("The heatmap show how the samples cluster, and it is useful to detect outliers. More informative is the MA plot which show the differentially expressed genes with an overview of the relation between fold-change and expression levels.\n")
  cat("\n", fill=T)

  plotFancyMA(res, base_name)

  res <- addGeneSymbol(res)
  res <- addWormbBaseAnnotations(res, annotation)
  res <- addTargetInfo(res, target_file)

  saveResults(res, base_name)

  save(
    dds, res, countdata_sub,
    file=paste0(
      "reports/",
      base_name,
      ".DE_DESeq2.RData")
    )
}
```


## 21U reads

```{r DifferentialExpression_21U, echo=F, results='asis', error=F, warning=F, message=F}
for (treatment in c("oxid", "untreated")){
  # subsection
  cat("\n", fill=T)
  cat("### QC", treatment, fill=T)  
  cat("\n", fill=T)
  # Prepare DESeq2 object
  smRNAclass <- "21U"
  countdata_sub <- selectSamples(countdata_m, c(smRNAclass, treatment))

  dds <- runDESeq(countdata_sub, "N2")
  col_data <- as.data.frame(colData(dds))

  cat("A brief summary of the samples analysed:\n")
  print(knitr::kable(col_data))

  res <- results(dds)

  # ==========================================================================
  # QC
  comparison <- gsub(" ", "_",
    gsub(".*: condition ", "", res@elementMetadata$description[2])
    )

  base_name <- paste(
    smRNAclass, treatment,
    comparison,
    sep="_")
  performQC(
    dds=dds,
    groups=c("condition","replicate", "treatment"),
    base_name=base_name
    )
  cat("\n", fill=T)
  cat("These plots show the relationships between samples.
    Useful to detect outliers.")
  cat("\n", fill=T)
  cat("Assuming the QC went well, we can can not look at the results.
    ")
  cat("\n", fill=T)

  #########################################################
  ## Results
  #########################################################
  cat("\n", fill=T)
  cat("### Results", treatment, fill=T)  
  cat("\n", fill=T)

  res <- results(dds)
  res <- res[order(res$padj),]
  capture.output(summary(res))
  up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
  up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
  down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
  down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

  sum(res$padj < 0.1, na.rm=TRUE)


  cat("Summary contains an overview of the results. There are", up, " significantly up-regulated genes and", down, "significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then",  up_log2, "are up-regulated and", down_log2, "are down-regulated.
  \n")
  cat("\n", fill=T)
  
  cat("The heatmap show how the samples cluster, and it is useful to detect outliers. More informative is the MA plot which show the differentially expressed genes with an overview of the relation between fold-change and expression levels.\n")
  cat("\n", fill=T)

  plotFancyMA(res, base_name)

  res <- addGeneSymbol(res)
  res <- addBiotype(res, annotation)
  res <- addTargetInfo(res, target_file)

  saveResults(res, base_name)

  save(
    dds, res, countdata_sub,
    file=paste0(
      "reports/",
      base_name,
      ".DE_DESeq2.RData")
    )
}
```


## 22G reads

```{r DifferentialExpression_22G, echo=F, results='asis', error=F, warning=F, message=F}
for (treatment in c("TAP")){
  # subsection
  cat("\n", fill=T)
  cat("### QC", treatment, fill=T)  
  cat("\n", fill=T)
  # Prepare DESeq2 object
  smRNAclass <- "22G"
  countdata_sub <- selectSamples(countdata_m, c(smRNAclass, treatment))

  dds <- runDESeq(countdata_sub, "N2")
  col_data <- as.data.frame(colData(dds))

  cat("A brief summary of the samples analysed:\n")
  print(knitr::kable(col_data))

  res <- results(dds)

  # ==========================================================================
  # QC
  comparison <- gsub(" ", "_",
    gsub(".*: condition ", "", res@elementMetadata$description[2])
    )

  base_name <- paste(
    smRNAclass, treatment,
    comparison,
    sep="_")
  performQC(
    dds=dds,
    groups=c("condition","replicate", "treatment"),
    base_name=base_name
    )
  cat("\n", fill=T)
  cat("These plots show the relationships between samples.
    Useful to detect outliers.")
  cat("\n", fill=T)
  cat("Assuming the QC went well, we can can not look at the results.
    ")
  cat("\n", fill=T)

  #########################################################
  ## Results
  #########################################################
  cat("\n", fill=T)
  cat("### Results", treatment, fill=T)  
  cat("\n", fill=T)

  res <- results(dds)
  res <- res[order(res$padj),]
  capture.output(summary(res))
  up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
  up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
  down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
  down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

  sum(res$padj < 0.1, na.rm=TRUE)


  cat("Summary contains an overview of the results. There are", up, " significantly up-regulated genes and", down, "significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then",  up_log2, "are up-regulated and", down_log2, "are down-regulated.
  \n")
  cat("\n", fill=T)
  
  cat("The heatmap show how the samples cluster, and it is useful to detect outliers. More informative is the MA plot which show the differentially expressed genes with an overview of the relation between fold-change and expression levels.\n")
  cat("\n", fill=T)

  plotFancyMA(res, base_name)

  res <- addGeneSymbol(res)
  res <- addBiotype(res, annotation)
  res <- addTargetInfo(res, target_file)

  saveResults(res, base_name)

  save(
    dds, res, countdata_sub,
    file=paste0(
      "reports/",
      base_name,
      ".DE_DESeq2.RData")
    )
}
```


## 26G reads

```{r DifferentialExpression_26G, echo=F, results='asis', error=F, warning=F, message=F}
for (treatment in c("oxid", "untreated")){
  # subsection
  cat("\n", fill=T)
  cat("### QC", treatment, fill=T)  
  cat("\n", fill=T)
  # Prepare DESeq2 object
  smRNAclass <- "26G"
  countdata_sub <- selectSamples(countdata_m, c(smRNAclass, treatment))


  if(treatment == "oxid"){
    cat("\n")
    cat("Replacing size factors")
    cat("\n")
    dds <- runDESeq(countdata_sub, "N2", size_factors=lib_sizes)
    col_data <- as.data.frame(colData(dds))
  }

  cat("A brief summary of the samples analysed:\n")
  print(knitr::kable(col_data))

  res <- results(dds)

  # ==========================================================================
  # QC
  comparison <- gsub(" ", "_",
    gsub(".*: condition ", "", res@elementMetadata$description[2])
    )

  base_name <- paste(
    smRNAclass, treatment,
    comparison,
    sep="_")
  performQC(
    dds=dds,
    groups=c("condition","replicate", "treatment"),
    base_name=base_name
    )
  cat("\n", fill=T)
  cat("These plots show the relationships between samples.
    Useful to detect outliers.")
  cat("\n", fill=T)
  cat("Assuming the QC went well, we can can not look at the results.
    ")
  cat("\n", fill=T)

  #########################################################
  ## Results
  #########################################################
  cat("\n", fill=T)
  cat("### Results", treatment, fill=T)  
  cat("\n", fill=T)

  res <- results(dds)
  res <- res[order(res$padj),]
  capture.output(summary(res))
  up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
  up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
  down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
  down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

  sum(res$padj < 0.1, na.rm=TRUE)


  cat("Summary contains an overview of the results. There are", up, " significantly up-regulated genes and", down, "significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then",  up_log2, "are up-regulated and", down_log2, "are down-regulated.
  \n")
  cat("\n", fill=T)
  
  cat("The heatmap show how the samples cluster, and it is useful to detect outliers. More informative is the MA plot which show the differentially expressed genes with an overview of the relation between fold-change and expression levels.\n")
  cat("\n", fill=T)

  plotFancyMA(res, base_name)

  res <- addGeneSymbol(res)
  res <- addBiotype(res, annotation)
  res <- addTargetInfo(res, target_file)

  saveResults(res, base_name)

  save(
    dds, res, countdata_sub,
    file=paste0(
      "reports/",
      base_name,
      ".DE_DESeq2.RData")
    )
}
```


```{r session}
sessionInfo()
Sys.info()
```