loadLibraries <- function(){
    libs=c(
    "data.table",
    "RColorBrewer",
    "ggplot2",
    "DESeq2",
    "scales",
    "pheatmap",
    "org.Ce.eg.db",
    "stringr"
  )
  lapply(libs, require, character.only=T)
}


# source("https://raw.githubusercontent.com/koundy/ggplot_theme_Publication/master/R/ggplot_theme_Publication.R")
theme_Publication <- function(base_size=14, base_family="Helvetica") {
      library(grid)
      library(ggthemes)
      (theme_foundation(base_size=base_size, base_family=base_family)
       + theme(plot.title = element_text(face = "bold",
                                         size = rel(1.2), hjust = 0.5),
               text = element_text(),
               panel.background = element_rect(colour = NA),
               plot.background = element_rect(colour = NA),
               panel.border = element_rect(colour = NA),
               axis.title = element_text(face = "bold",size = rel(1)),
               axis.title.y = element_text(angle=90,vjust =2),
               axis.title.x = element_text(vjust = -0.2),
               axis.text = element_text(), 
               axis.line.x = element_line(colour="black"),
               axis.line.y = element_line(colour="black"),
               axis.ticks = element_line(),
               panel.grid.major = element_line(colour="#f0f0f0"),
               panel.grid.minor = element_blank(),
               legend.key = element_rect(colour = NA),
               legend.position = "bottom",
               legend.direction = "horizontal",
               legend.key.size= unit(0.2, "cm"),
               legend.spacing = unit(0, "cm"),
               legend.title = element_text(face="italic"),
               plot.margin=unit(c(10,5,5,5),"mm"),
               strip.background=element_rect(colour="#f0f0f0",fill="#f0f0f0"),
               strip.text = element_text(face="bold")
       ))
      
}


scale_fill_Publication <- function(...){
      library(scales)
      discrete_scale("fill","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)
      
}


scale_colour_Publication <- function(...){
      library(scales)
      discrete_scale("colour","Publication",manual_pal(values = c("#386cb0","#fdb462","#7fc97f","#ef3b2c","#662506","#a6cee3","#fb9a99","#984ea3","#ffff33")), ...)
      
}


concat <- function(...) paste(..., sep="")

df2matrix <- function(df){
# converts a dataframe in matrix
# using the 1st column as rownnames
  # # test
  # df=countdata
  mat <- as.matrix(df[, -1])
  rownames(mat) <- df[, 1]
  class(mat) <- "numeric"
  return(mat)
}


##
## @brief      { function_description }
##
## @param      dir   The dir
##
## @return     { description_of_the_return_value }
##
prepareCountMatrix <- function(dir="./"){
  # reads files generated by htseq-counts
  # creates a tmp DESeqDataSet
  # returns the count matrix
  sample_files <- grep(".counts", list.files(dir), value=TRUE)
  sample_name <- gsub('.counts', "", sample_files)

  sample_name  <- ifelse(
    str_detect(sample_name, '\\.\\d+'),
    sample_name,
    paste0(sample_name, ".All")
  )

  conds <- gsub(
     "(^\\w+)_\\w+_r\\d+_\\w+.\\w+.counts", "\\1",
     sample_files
     )


  sample_table <- data.frame(
    sampleName = sample_name,
    fileName = sample_files,
    condition = sample_name)

  dds <- DESeqDataSetFromHTSeqCount(
    sampleTable = sample_table,
    directory = dir,
    design= ~ condition)

  countdata_m <- counts(dds)
  return(countdata_m)
}


createBiotypesTable <- function(mat){
  library("biomaRt")
  
  tab <- data.frame(wormbase_gene = rownames(mat))
  wormbase <- useMart(biomart = "parasite_mart", host = "parasite.wormbase.org")
  wormbase <- useDataset(mart = wormbase, dataset = "wbps_gene")
  bio <- getBM(
    attributes = c("wbps_gene_id", "gene_biotype"), 
    filters="wbps_gene_id", 
    values=tab$wormbase_gene, 
    mart=wormbase)
  idx <- match(tab$wormbase_gene, bio$wbps_gene_id)
  tab$biotype <- bio$gene_biotype[ idx ]
  return(tab)
}


excludeNonStructuralReads <- function(
  mat,
  annotation,
  type=c("df", "file")
  ){
  if(type == "file"){
    bio <- read.table(annotation,
       sep="\t",
       header=FALSE,
       stringsAsFactors=FALSE
    )
  } else {
    bio <- annotation
  }
  colnames(bio) <- c("wormbase_gene", "gene_biotype")
  exclude_bio <- c("miRNA", "piRNA", "rRNA", "tRNA", "snoRNA", "snRNA")
  row_keep <- subset(bio, !(bio$gene_biotype %in% exclude_bio))$wormbase_gene
  mat <- mat[ rownames(mat) %in% row_keep, ]
  return(mat)
}


selectSamples <- function(mat, sample_names){
  # given a vector with conditions, will select cols that match
  if (length(sample_names) > 1){
    # only those with both strings:
    cols <- sapply(sample_names, function(x) grep(x, colnames(mat)))
    mat <- mat[, Reduce(intersect, cols)]
  } else {
    mat <- mat[, grepl(sample_names, colnames(mat))]
  }
  colnames(mat) <- gsub(".\\w+$", "", colnames(mat))
  return(mat)
}


runDESeq <- function(mat=countdata_sub, control="N2", size_factors=NULL){
  # Takes a matrix of counts and prepares the DESeq2 obj
  # control is there to make sure the comparison is in the
  # correct direction, that is, condition vs control and not
  # control vs condition 
  # Sample, condition names, etc are taken from the name and assume 
  # C. elegans pipeline format
  # returns DESeq2 object
  samples <- colnames(mat)
  conds <- factor(
    stringr::str_sub(colnames(mat), 1,2)
  )

  replicates <- factor(
    gsub(
     "^\\w+_\\w+_(r\\d+)_\\w+.\\w+", "\\1",
     colnames(mat)
     )
  )

  treatment <- factor(
    gsub(
     "^\\w+_\\w+_r\\d+_(\\w+)", "\\1",
     colnames(mat)
     )
  )

  coldata <- data.frame(
    condition=conds,
    replicate=replicates,
    treatment=treatment,
    row.names=samples
  )


  ## create deseq object will all the data
  # in more fine-tuned analysis factors can be subseted

  dds <- DESeqDataSetFromMatrix(
    countData = mat,
    colData = coldata,
    design = ~ condition
  )
  # making sure that N2 is the first level, and thus
  # log2FC are reported as treatment over control (xf/N2)
  dds$condition <- relevel(dds$condition, control)

  if(!(is.null(size_factors))){
    sizeFactors(dds) <- size_factors
  }

  ## DESeq
  dds <- DESeq(dds)
  col_data <- as.data.frame(colData(dds))
  # print(col_data)
  return(dds)
}


performQC <- function(dds=dds, groups, base_name, dir="reports/figure/"){
  # this function will produce QC plots
  # Inputs are the dds and a vector of strings with
  # conditions to be used for visualization
  # counts
  select <- order(rowMeans(counts(dds,normalized=TRUE)),decreasing=TRUE)[1:20]
  nt <- normTransform(dds) # defaults to log2(x+1)
  log2.norm.counts <- assay(nt)[select,]
  df <- as.data.frame(colData(dds)[, groups])

  rld <- rlog(dds, blind=FALSE)
  vsd <- varianceStabilizingTransformation(dds, blind=FALSE)
  # sample clustering
  sampleDists <- dist(t(assay(rld)))
  sampleDistMatrix <- as.matrix(sampleDists)
  rownames(sampleDistMatrix) <- paste(rld$condition, rld$replicate, rld$treatment, sep="-")
  colnames(sampleDistMatrix) <- NULL

  colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
  pheatmap(sampleDistMatrix,
     clustering_distance_rows=sampleDists,
     clustering_distance_cols=sampleDists,
     col=colors)
  cat("\n", fill=T)
  # pdf(paste0(dir, base_name, ".pdf"))
  plotDispEsts(dds)
  cat("\n", fill=T)
  pheatmap(
     log2.norm.counts,
     cluster_rows=FALSE,
     show_rownames=TRUE,
     cluster_cols=TRUE,
     annotation_col=df
     )
  cat("\n", fill=T)
  nudge <- position_nudge(y = 1.5)
  # p1 <- plotPCA(rld, intgroup=c("treatment")) +
  #    ggtitle(smRNAclass) +
  #    geom_label(aes(label = name), 
  #     position = nudge, 
  #     size = 3)
  # print(p1)

  data <- plotPCA(rld, intgroup=groups, returnData=TRUE)
  percentVar <- round(100 * attr(data, "percentVar"))
  p2 <- ggplot(data, aes(PC1, PC2, color=condition, shape=treatment)) +
     geom_point(size=3) +
     geom_label(aes(label = name), 
      position = nudge, 
      size = 3,
      vjust = "inward",
      hjust = "inward") +
     ggtitle(base_name) +
     xlab(paste0("PC1: ",percentVar[1],"% variance")) +
     ylab(paste0("PC2: ",percentVar[2],"% variance")) 
  cat("\n", fill=T)
  print(p2)
  cat("\n", fill=T)
  # dev.off()
}


plotFancyMA <- function(res, base_name, dir="reports/figure/", fdr=0.01){
  up <- sum(res$padj < fdr & res$log2FoldChange > 0, na.rm=TRUE)
  down <- sum(res$padj < fdr & res$log2FoldChange < 0, na.rm=TRUE)
  x_pos <- max(res@listData$baseMean) * 0.7
  tit <- paste("Differentially expressed genes\nFDR =", fdr)
  pdf(paste0(dir, base_name, ".MAplot.pdf"))
  plotMA(res, main=tit, ylim=c(-5,5))
  text(x=x_pos, y=2, labels=up, col="red", cex=1.5)
  text(x=x_pos, y=-2, labels=down, col="red", cex=1.5)
  dev.off()
  plotMA(res, main="DESeq2", ylim=c(-5,5))
  text(x=x_pos, y=2, labels=up, col="red", cex=1.5)
  text(x=x_pos, y=-2, labels=down, col="red", cex=1.5)
}


addGeneSymbol <- function(res, orgDB=org.Ce.eg.db){
  res$wormbase_gene <-rownames(res)

  res$symbol <- mapIds(orgDB,
               keys=row.names(res),
               column="SYMBOL",
               keytype="WORMBASE",
               multiVals="first")
  res$type <- mapIds(orgDB,
               keys=row.names(res),
               column="GENENAME",
               keytype="WORMBASE",
               multiVals="first")

  return(res)
}


addBiotype <- function(res, annotation){
  bio <- read.table(annotation,
     sep="\t",
     header=T,
     stringsAsFactors=FALSE
  )
  
  idx <- match(res$wormbase_gene, bio$wormbase_gene)
  res$gene_biotype <- bio$gene_biotype[ idx ]
  return(res)
}

addTransposonInfo <- function(res, annotation){
  if(length(grep("wormbase_gene", colnames(res))) == 0) {
      res$wormbase_gene <-rownames(res)
    } 
  bio <- read.table(annotation,
     sep="\t",
     header=T,
     stringsAsFactors=FALSE
  )
  
  idx <- match(res$wormbase_gene, bio$WB.Transposon.ID)
  res$wormbase_gseq <- bio$Sequence.parent[ idx ]
  res$Family <- bio$Family[ idx ]
  # res$Remark <- bio$Remark[ idx ]
  return(res)
}

addTargetInfo <- function(res, target_file){
  targets <- read.delim(target_file, header = TRUE, stringsAsFactors=FALSE)
  targets <- aggregate(targets[2], targets[-2],
     FUN = function(X) paste(unique(X), collapse=",")
     )
  res$Target_gene <- ifelse(
    res$symbol %in% targets$geneid,
    targets$exp,
    "unknown")
  return(res)
}


addWormbBaseAnnotations <- function(res){
  library("biomaRt")
  
  res <- data.frame(res)  
  if(length(grep("wormbase_gene", colnames(res))) == 0) {
      res$wormbase_gene <-rownames(res)
    } 
    
  wormbase <- useMart(biomart = "parasite_mart", host = "parasite.wormbase.org")
  wormbase <- useDataset(mart = wormbase, dataset = "wbps_gene")
  bio <- getBM(
    attributes = c("wbps_gene_id", "wormbase_gseq", "external_gene_id", "gene_biotype"), 
    filters="wbps_gene_id", 
    values=res$wormbase_gene, 
    mart=wormbase)
  # setDT(bio)
  # merge(res, bio, by.x="wormbase_gene", by.y="wbps_gene_id", all.x=TRUE)
  idx <- match(res$wormbase_gene, bio$wbps_gene_id)
  res$gene_id <- bio$external_gene_id[ idx ]
  res$wormbase_gseq <- bio$wormbase_gseq[ idx ]
  res$biotype <- bio$gene_biotype[ idx ]
  return(res)
}


saveResults <- function(res, base_name, dir="reports/") {
  res_out <- as.data.frame(res)
  res_out <- data.frame(
  wormbase_gene = res_out[ , names(res_out) == "wormbase_gene"],
  res_out[ , !(names(res_out) == "wormbase_gene")])

  write.table(
    res,
  file=paste0(dir, base_name, ".DESeq2_results.txt"),
  sep="\t",
  quote=FALSE,
  row.names=FALSE,
  col.names=TRUE)
}
