---
title: "Differential gene expression of small RNA-seq in C. elegans"
author: "Ant√≥nio Domingues"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_float: true
    theme: readable
    highlight: tango
---


```{r setchunk, cache=FALSE, include=FALSE}
knitr::opts_chunk$set(
  include=TRUE,
  echo = FALSE,
  tidy=TRUE,
  fig.path = "reports/figure/",
  dev=c("png","pdf"),
  fig.align="center")
```


```{r setup, results="hide", include=FALSE}
source("DEG_analysis_DESeq2.helpers.R")

loadLibraries()

lapply(
  c("reports", "reports/figure"),
  function(x) dir.create(file.path(x), showWarnings = FALSE)
)

```


```{r input data}
transposon_annotation <- "/home/adomingu/imb-kettinggr/genomes/Caenorhabditis_elegans/Ensembl/WBcel235/Annotation/Genes/WS255_transposon_info.txt"

directory <- "./"
```


# Read count

```{r prepare_counts, results="hide", include=FALSE}
countdata <- prepareCountMatrix()
countdata_m <- countdata
dim(countdata_m)

```

```{r select_conditions, results="hide", include=FALSE}
smRNAclass <- "All"
base_name <- "all_reads_and_treatments"
countdata_sub <- selectSamples(countdata_m, c(smRNAclass))
dim(countdata_sub)
dds <- runDESeq(countdata_sub, "N2")
treatments <- unique(as.character(colData(dds)$treatment))
knitr::kable(colData(dds))
```


## QC

### Dispersion

```{r dispersion_plot}
performQC(dds, c("condition","replicate", "treatment"), base_name)
```

### Sample clustering

Here we extract the counts and transformed values to use for clustering and visualization:
- The function rlog, stands for regularized log, transforming the original count data to the log2 scale by fitting a model with a term for each sample and a prior distribution on the coefficients which is estimated from the data.
-  Variance stabilizing transformation (vsd) is a parametric fit for the dispersion.


Heatmaps showing the expression data of the 20 most highly expressed genes. The data is of log2 normalized counts, from regularized log transformation  and from variance stabilizing transformation.

Now we can actually cluster the samples:


# Results

There are many classes of _C. elegans_ small RNAs which can be enriched with different treatments. Small RNAs are split into classes according to size and nucleotide bias. For the differential expression of **Transposons** we will focus on the the TAP or Rpph treatments, that enrich for 22Gs, and the full libraries or 22G reads. From the [Wormbook](http://www.wormbook.org/chapters/www_endoRNAipathwys/endoRNAipathwys.html#bib52):

> The WAGO 22G RNAs bind semi-redundant Argonautes of the worm-specific WAGO clade to mediate silencing of certain protein-coding genes, __transposons__, pseudogenes, and cryptic loci through both transcriptional and post-transcriptional mechanisms (Guang et al., 2008; Gu et al., 2009; Burkhart et al., 2011; Burton et al., 2011; Ashe et al., 2012; Buckley et al., 2012; Gu, S.G., et al. 2012; Luteijn et al., 2012; Shirayama et al., 2012). 

[and](http://www.wormbook.org/chapters/www_endoRNAipathwys/endoRNAipathwys.html#sec10):

> Consistent with their critical, non-redundant role in the WAGO 22G RNA pathway that silences __transposons__, loss of any of these proteins results in mutator (Mut), ts sterile, Him, and Rde phenotypes (Collins et al., 1987 Ketting et al., 1999; Tijsterman et al., 2002a; Vastenhouw et al., 2003; Chen et al., 2005; Tops et al., 2005; Gu et al., 2009; Phillips et al., 2012), although a mut-14 mutant carrying a missense mutation shows only a germline Rde phenotype (Tijsterman et al., 2002a; Phillips et al., 2012).


## 22G reads

```{r }
cl <-  c("TAP", "rpph")
n_conds <- length(
  grep(
    paste(cl, collapse="|"),
    treatments,
    ignore.case = TRUE)
  )
if (n_conds < 1){
  do_something <- FALSE
  cat("\n", fill=T)
  cat("The library treatments in this experiment are not adequate for an analysis of this RNA class.", paste(cl, collapse=" or "), "were expected.", fill=T)
  cat("\n", fill=T)
} else {
  do_something <- TRUE
}

```


```{r DifferentialExpression_22G, eval=do_something, echo=F, results='asis', error=F, warning=F, message=F}
treat <- treatments[grepl(
  paste(c("TAP", "Rpph"), collapse="|"),
  treatments,
  ignore.case = TRUE)]

for (treatment in treat){
  # subsection
  cat("\n", fill=T)
  cat("### QC", treatment, fill=T)  
  cat("\n", fill=T)
  # Prepare DESeq2 object
  smRNAclass <- "22G"
  countdata_sub <- selectSamples(countdata_m, c(smRNAclass, treatment))

  dds <- runDESeq(countdata_sub, "N2")
  col_data <- as.data.frame(colData(dds))

  cat("A brief summary of the samples analysed:", fill=T)
  cat("\n", fill=T)
  print(knitr::kable(col_data))
  cat("\n", fill=T)
  res <- results(dds)

  # ==========================================================================
  # QC
  comparison <- gsub(" ", "_",
    gsub(".*: condition ", "", res@elementMetadata$description[2])
    )

  base_name <- paste(
    smRNAclass, treatment,
    comparison,
    sep="_")
  performQC(
    dds=dds,
    groups=c("condition","replicate", "treatment"),
    base_name=base_name
    )
  cat("\n", fill=T)
  cat("These plots show the relationships between samples.
    Useful to detect outliers.")
  cat("\n", fill=T)
  cat("Assuming the QC went well, we can can not look at the results.
    ")
  cat("\n", fill=T)

  #########################################################
  ## Results
  #########################################################
  cat("\n", fill=T)
  cat("### Results", treatment, fill=T)  
  cat("\n", fill=T)

  res <- results(dds)
  res <- res[order(res$padj),]
  capture.output(summary(res))
  up <- sum(res$padj < 0.1 & res$log2FoldChange > 0, na.rm=TRUE)
  up_log2 <- sum(res$padj < 0.1 & res$log2FoldChange > 1, na.rm=TRUE)
  down <- sum(res$padj < 0.1 & res$log2FoldChange < 0, na.rm=TRUE)
  down_log2 <- sum(res$padj < 0.1 & res$log2FoldChange < -1, na.rm=TRUE)

  sum(res$padj < 0.1, na.rm=TRUE)


  cat("Summary contains an overview of the results. There are", up, " significantly up-regulated genes and", down, "significantly down-regulated genes. If we consider only genes with more than a 2 fold change, then",  up_log2, "are up-regulated and", down_log2, "are down-regulated.
  \n")
  cat("\n", fill=T)
  
  cat("The heatmap show how the samples cluster, and it is useful to detect outliers. More informative is the MA plot which show the differentially expressed genes with an overview of the relation between fold-change and expression levels.\n")
  cat("\n", fill=T)

  plotFancyMA(res, base_name)

  res <- addTransposonInfo(res, transposon_annotation)

  saveResults(res, base_name)

  save(
    dds, res, countdata_sub,
    file=paste0(
      "reports/",
      base_name,
      ".DE_DESeq2.RData")
    )
}
```

# Session info

```{r session}
sessionInfo()
Sys.info()
```