---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: true
    toc_float: true
    css: styles.css
---

<div class="contentbox">

```{r Initialise, echo=F,result='hide',error=F,warning=F,message=F}
require(rmarkdown)
require(knitr)
require(Cairo)                                                    
knitr::opts_chunk[['set']](dev='CairoPNG')
source('ChIP.shinyrep.helpers.R')
loadGlobalVars()
targets <<- ChIPhelper.init("readTargets")
peaks   <<- ChIPhelper.init("readPeaks")
``` 


## Raw reads QC ##

The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

1. The "Duplication levels" plots show the proportion of the library which is made up of sequences in each of the different duplication level bins. It is useful for the assessment of the NGS library complexity and the presence or absence of PCR over-amplification during library prep. There are two lines on the plot. The blue line takes the full sequence set and shows how its duplication levels are distributed. In the red plot the sequences are de-duplicated and the proportions shown are the proportions of the deduplicated set which come from different duplication levels in the original data. In a properly diverse library most sequences should fall into the far left of the plot in both the red and blue lines. A general level of enrichment, indicating broad oversequencing in the library will tend to flatten the lines, lowering the low end and generally raising other categories. The presence of dominant sequences (e.g. adapter dimers) will tend to produce spikes towards the right of the plot. These high duplication peaks will most often appear in the red trace as they make up a high proportion of the original library, but usually disappear in the blue trace as they make up an insignificant proportion of the deduplicated set. If peaks persist in the blue trace then this suggests that there are a large number of different highly duplicated sequences which might indicate either a contaminant set or a severe PCR duplication. See the related PBC metrics below.

2. The "Read qualities" Box-Whisker plots show the range of quality values across all base positions:
    (i) The central red line is the median value,
    (ii) The yellow box represents the inter-quartile range (25-75%),
    (iii) The upper and lower whiskers represent the 10% and 90% points,
    (iv) The blue line represents the mean quality.
The y-axis on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area.

3. The "Sequence bias" plots show the proportion of each base (% G, A, T and C) at each position. In a typical ChIP-seq library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other, G being similar to C and A to T. 

```{r FastQC_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.Fastqc(web=F),sep="\n")
```

## Read mapping ##

Mapping to the reference genome was done using the established NGS mapping tool Bowtie (http://bowtie-bio.sourceforge.net/index.shtml). The mapping statistics show the total number of reads delivered to the aligner ("all reads"), the number of uniquely mapping reads ("mapped"), the reads that failed to align ("unmapped"), the number of reads discarded because of mapping to multiple positions ("too many map. pos.") and the number of duplicate reads ("duplicates") amongst the uniquely mapped reads. The browser tracks are generated using bamCoverage tool in [deepTools](https://deeptools.readthedocs.io/en/develop/) and they are normalized using the Counts Per Million mapped reads (CPM) method. The CPM formula is: CPM  = number of reads per bin (10 bp) / number of mapped reads (in millions).

```{r Bowtie_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.Bowtie(),sep="\n")
```

## UCSC Genome Browser tracks ##

To view the tracks in the UCSC Genome Browser, navigate to the [UCSC Genome Browser Track Hub page](https://genome-euro.ucsc.edu/cgi-bin/hgHubConnect) and enter the following line into the `URL` field:

```{r Trackhub, echo=F, results='asis', error=F, warning=F, message=F}
cat(ChIPhelper.Trackhub(), sep="\n")
```

## PCR bottleneck coefficient ##

The PBC (PCR bottleneck coefficient) is an approximate measure of library complexity. Provisionally, 0-0.5 is severe bottlenecking, 0.5-0.8 is moderate bottlenecking, 0.8-0.9 is mild bottlenecking, while 0.9-1.0 is no bottlenecking.

Low values can indicate a technical problem, such as library complexity bottleneck. Nuclease-based assays (DNase, MNase) detecting features with base-pair resolution (transcription factor footprints, positioned nucleosomes) are expected to recover the same read multiple times, resulting in a lower PBC score for these assays. Note that the most complex library, random DNA, would approach 1.0, thus the very highest values can indicate technical problems with libraries. Some common numbers from ENCODE datasets are: 82% for TF ChIPs, 89% for Histone ChIPs, 77% for DNase, 98% for FAIRE, and 97% for control ENCODE datasets with no or mild bottlenecking.

```{r PBC_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.PBC(),sep="\n")
```

## IPstrength ##

To estimate the IP strength, we attempt to decompose the population of IP reads into two distinct components: those pulled down by the antibody, and background. The enrichment plot is useful to detect several forms of biases:

* The distance between the two curves depends on the enrichment of the IP versus the input; close curves suggest weak or diffuse enrichment.
* Under-sequencing produces curves close to the x-axis for a large percentage of bins since no reads were aligned on this bins. 
* Sequencing bias towards a small set of specific regions produces curves shifted towards the y-axis for a large percentage of bins.

For more details see Fig.2 of the CHANCE paper (https://doi.org/10.1186/gb-2012-13-10-r98).

```{r IPstrength_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.IPstrength(web=F),sep="\n")
```

## Cross-correlation analysis ##

**Cross-correlation**
A measure of enrichment derived without dependence on prior determination of enriched regions. Forward and reverse strand read coverage signal tracks are computed (number of unique mapping read starts at each base in the genome on the + and - strand counted separately). The forward and reverse tracks are shifted towards and away from each other by incremental distances and for each shift, the Pearson correlation coefficient is computed. In this way, a cross-correlation profile is computed, representing the correlation between forward and reverse strand coverage at different shifts. The highest cross-correlation value is obtained at a strand shift equal to the predominant fragment length in the dataset as a result of clustering/enrichment of relative fixed-size fragments around the binding sites of the target factor or feature.

**Normalized Strand Cross-correlation coefficient (NSC):**
The NSC is the ratio of the maximal cross-correlation value (which occurs at strand shift equal to fragment length) divided by the background cross-correlation (minimum cross-correlation value over all possible strand shifts). Higher values indicate more enrichment, values less than 1.1 are relatively low NSC scores, and the minimum possible value is 1 (no enrichment). This score is sensitive to technical effects; for example, high-quality antibodies such as H3K4me3 and CTCF score well for all cell types and ENCODE production groups, and variation in enrichment in particular IPs is detected as stochastic variation. This score is also sensitive to biological effects; narrow marks score higher than broad marks (H3K4me3 vs H3K36me3, H3K27me3) for all cell types and ENCODE production groups, and features present in some individual cells, but not others, in a population are expected to have lower scores.

**Relative Strand Cross-correlation coefficient (RSC):**
The RSC is the ratio of the fragment-length cross-correlation value minus the background cross-correlation value, divided by the phantom-peak cross-correlation value minus the background cross-correlation value. The minimum possible value is 0 (no signal), highly enriched experiments have values greater than 1.

```{r PhantomPeak_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.PhantomPeak(web=F),sep="\n")
```

## ChIP-seq peaks ##


The ENCODE consortium empirically identified 'blacklist regions' (BRs) for some species and genome versions. These blacklist regions are often found at specific types of repeats (eg. centromeres, telomeres and satellite repeats) and are troublesome for high throughput sequencing aligners. These regions also confuse peak callers and result in spurious signal. Therefore, it is suggested to check the amount of peaks overlapping these regions and remove them for the downstream analysis. The table shows total number of peaks detected by MACS2, the number of peaks without overlapping the blacklist regions and the percentage of peaks overlapping these regions.  

Note that these blacklists are applicable to human (hg38 and hg19), mouse (mm10 and mm9), worm (ce10) and fly (dm3) genomes.


```{r BLACKLISTFILTER_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(ChIPhelper.BLACKLISTFILTER(),sep="\n")
```

The number of peaks called at the default false-discovery rate (FDR) cutoff of 5% by MACS2 (https://github.com/taoliu/MACS/) in the different comparisons is shown below.

```{r PeakCall_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
peaks <- ChIPhelper.init("readPeaks")

l <- lapply(peaks,function(x) {
    x <- table(x$chr)
    xx <- as.numeric(x)
    names(xx) <- names(x)
    xx
})
p <- Reduce(function(x,y) {
    z <- merge(x,y,all=T,by=0)
    rownames(z) <- z[,1]
    z[,-1]
},l)
#colnames(p) <- 1:ncol(p)
colnames(p) <- names(l)

# print the legend
#cat("",fill=T)
#cat(kable(data.frame(column=1:length(l),sample=names(l))),sep="\n")
#cat("",fill=T)

# print the peaks table
cat("",fill=T)
cat(kable(p,row.names=T),sep="\n")
cat("",fill=T)
```

## Correlation of peaks ##

The following Venn diagrams correlate the peaks called in the sample groups.

```{r peak_venn_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
ChIPhelper.VennDiagram()
```

## Functional annotation of peaks ##

The following plot show the feature distribution based on their chromosome region 

### Feature distribution of peaks for all samples ###

```{r feature_distribution_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
functionalPlot <- paste0( SHINYREPS_PEAK_ANNOTATION, "/", "ChIPseq_Feature_Distribution_Barplot.png")
cat("![Peak_Annotation img](", functionalPlot, ")",fill=T)
```

### Distribution of TF-binding loci relative to TSS for all samples ### 

```{r feature_distribution_TSS_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
    functionalPlotTSS <- paste0( SHINYREPS_PEAK_ANNOTATION, "/", "ChIPseq_Feature_Distribution_Related_to_TSS_Barplot.png")
    cat("![Peak_Annotation img](", functionalPlotTSS, ")",fill=T)
```

### ChIP peaks coverage over Chromosomes ###  

```{r peak_annotation_coverage_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
    cat(ChIPhelper.peakAnnotationCoverage(web=F),sep="\n")
```

### UpSet Plot ###

```{r peak_annotation_upset_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
    cat(ChIPhelper.peakAnnotationUpSet(web=F),sep="\n")
```

## Enrichment Analysis ##

Enrichment Analysis is carried out using GREAT (http://great.stanford.edu). It assigns each gene a regulatory domain consisting of basal domain that extends 5 kb upstream and 1 kb downstream from its transcription start site, and and an extension up to the basal regulatory domain of the nearest upstream and downstream genes within 1 Mb. GREAT performs both the binomial test over genomic regions and the hypergeometric test over genes to provide an accurate picture of annotation enrichments for genomic regions. The ontology terms reported below are ranked by the binomial FDR q-value.   

Note that GREAT supports recent assemblies of the well-annotated human (UCSC hg19 and UCSC hg18), mouse (UCSC mm9) and zebrafish (UCSC danRer7) genomes.

```{r Enrichment_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center"}

for(i in 1:length(levels(targets$group))) {

    # subsection
    cat("### ", levels(targets$group)[i], fill=T)
    cat("", fill=T)

    # GO Enrichment Plot

    cat("#### GO terms enriched (Biological Process, BP)", fill=T)

    GO_bar     <- paste0(SHINYREPS_GREAT, "/", gsub("=.*", "", levels(targets$group)[i]),"_GO_Barplot.png")
    GO_table   <- paste0(SHINYREPS_GREAT, "/", gsub("=.*", "", levels(targets$group)[i]), "_GO_Enrichment.csv")
    
    # if there are GO enrichment results (the barplot was created)
    if(file.exists(GO_bar)) {

        # GO enriched terms in a barplot
        cat("Listed are the top five enriched GO terms (BP) found using peaks with FDR < 0.01 detected in all replicates.", fill=T)
        cat("![GREAT_analysis img](", GO_bar, ")\n", fill=T)
    
        # and table of GO enriched terms
        cat("The table shows the top 10 enriched GO terms (BP):\n", fill=T)
        enrich  <- read.csv(file=GO_table, row.names=1)
        r       <- min(10, nrow(enrich))
        enrich  <- as.data.frame(enrich)
        enrich  <- cbind(enrich[1], enrich[2], enrich[8], enrich[9], enrich[6], enrich[5], enrich[7])
        colnames(enrich) <- c("GO ID","Term Name", "Binom Raw P-value", "Binom FDR Q-Val", "Binom Fold Enrichment", "Binom Observed Region Hits", "Binom Region Set Coverage" )
        #colnames(enrich)  <- gsub("GO.Biological.Process.", "",  colnames(enrich))
        #colnames(enrich)  <- gsub("\\.", " ", colnames(enrich))
        cat(kable(enrich[1:r,], row.names=T, align="r"),  sep="\n") # show only the top min or top 10 biological processes

        cat("\n", fill=T)

    } else {
        cat("no GO enrichment results available\n", fill=T)
    }



    # Pathway Enrichment Plot
    # up-regulated genes
    cat("#### Pathways enriched (Molecular Signatures Database, MSigDB)", fill=T)

    Pathway_bar     <- paste0(SHINYREPS_GREAT, "/", gsub("=.*", "", levels(targets$group)[i]), "_MSigDB_Pathway_Barplot.png")
    Pathway_table   <- paste0(SHINYREPS_GREAT, "/", gsub("=.*", "", levels(targets$group)[i]), "_MSigDB_Pathway_Enrichment.csv")
   
    # if there are Pathway enrichment results (the barplot was created)
    if(file.exists(Pathway_bar)) {

        # Pathways enriched in a barplot
        cat("Listed are the top five enriched MSigDB Pathways found using peaks with FDR < 0.01 detected in all replicates.", fill=T)
        cat("![GREAT_analysis img](", Pathway_bar, ")\n", fill=T)
    
        # and table of Pathways enriched
        cat("The table shows the top 10 enriched MSigDB Pathways:\n", fill=T)
        enrich  <- read.csv(file=Pathway_table, row.names=1)
        r       <- min(10, nrow(enrich))
        enrich  <- as.data.frame(enrich)
        enrich$MSigDB.Pathway.ID <- ifelse(nchar(as.character(enrich$MSigDB.Pathway.ID)) > 10, paste0(substr(enrich$MSigDB.Pathway.ID, 0, 15), "..."), enrich$MSigDB.Pathway.ID) # shorten long lines
        enrich  <- cbind(enrich[1], enrich[2], enrich[8], enrich[9], enrich[6], enrich[5], enrich[7])
        colnames(enrich) <- c("Pathway ID","Term Name", "Binom Raw P-value", "Binom FDR Q-Val", "Binom Fold Enrichment", "Binom Observed Region Hits", "Binom Region Set Coverage" )
        #colnames(enrich)  <- gsub("\\.", " ", colnames(enrich))
        cat(kable(enrich[1:r,], row.names=T, align="r", format="markdown"), sep="\n") # show only the top min or top 10 pathways
        cat("\n", fill=T)

    } else {
        cat("no Pathway enrichment results available\n", fill=T)
    }
}
```

## Differential binding analysis {.tabset}

```{r diffbind, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center"}
invisible(ChIPhelper.diffbind())
```

## Used tools and versions for this analysis ##
```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

</div>
