\documentclass[a4paper,10pt]{article}
\usepackage[utf8]{inputenc}
\usepackage{graphicx}
\usepackage{amsmath}
\usepackage{tabularx}
\usepackage{multicol}
\usepackage{caption}
\usepackage{array}
\usepackage{tabularx}
\usepackage{rotating}
\usepackage{longtable}	%for FASTQC.tex --> splits big tables into multiple pages
\usepackage[a4paper,hmargin=2cm,vmargin=2cm]{geometry}%margins

%%
%% Title Page
%%
\title{imb\_soshnikova\_2014\_03\_soshnikova\_MBDseq}
\author{Sergi Sayols Puig}
\date{April 17, 2014}
\begin{document}
\maketitle

%%
%% QC
%%
\section{Quality Control RNAseq}

%% SAV
\subsection{Sequencing quality}
The sequencing quality of the run was good, and the read distribution over the libraries was good. See tables \ref{Chip Summary}, \ref{Lane Results Summary: Read 1} and \ref{Expanded Lane Summary: Read 1} for details. 

\input{BustardSummary.tex}

%% Bowtie
\subsection{Mapping}
Mapping was done using Bowtie version 1.0.0 against the current mm9 the assembly version. The following parameters were used to get only uniquely mapping reads:

{\tiny
\begin{table}[h!]
\begin{tabular}{cll}\hline
	parm & value & desc\\\hline
	-q & & reads contain qualities\\
	--sam  & & output in standard sam format\\
	--best & & report best alignment\\
	--strata & & from the best startum (less number of mismatches)\\
	--nomaqround & & do not round qualities\\
	-l & 51 & seed length of the read\\
	-n & 3 & allow up to 3 mismatches in the seed\\
	-e & 120 & the sum of the qualities of all mismatche smust not exceed 120\\
	-m & 1 & supress reads mapping with more than 1 reportable alignments\\
	-p & 16 & use 16 threads \\\hline
\end{tabular}
\end{table}}

Overall stats are displayed in table \ref{mapping summary}.
\input{bowtie+dupmetrics.tex}

%\clearpage

%% FASTQC report
\subsection{Raw reads qualities, sequence bias and duplication}

\input{FASTQC.tex}

%% PCA report generated by the DESeq program with the 25 most variant miRNAs.
\subsection{Batch effect}
PCA report by splitting the genome into 1000bp long bins, and counting the reads ammped on them. The first figure shows how samples cluster by taking the 500 most variant bins. The second figure shows raw Euclidean distances between samples by taking bins with variance higher 10. The third figure shows the first two components of a PCA by using these bins.

{\tiny
\begin{longtable}{@{}ccc@{}}
\centering
	\includegraphics[page=1,width=.33\textwidth]{{../qc/mbd_binned_pca}.pdf} &
	\includegraphics[page=2,width=.33\textwidth]{{../qc/mbd_binned_pca}.pdf} &
	\includegraphics[page=3,width=.33\textwidth]{{../qc/mbd_binned_pca}.pdf} \\
\caption{Batch effect.}\label{Batch}
\end{longtable}}
%\clearpage

%%
%% BIBLIOGRAPHY
%%
%\nocite{*}	%display the complete biblio without previous citation
%\small
%\bibliographystyle{unsrt}
%\bibliography{rnaseq}

\end{document}
