---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: true
    toc_float: true
    css: styles.css
---

<div class="contentbox">

```{r Initialise, echo=F, result='hide', error=F, warning=F, message=F}
options(stringsAsFactors=FALSE)
require(rmarkdown)
require(knitr)
require(Cairo)                                                    
knitr::opts_chunk[['set']](dev='CairoPNG')
source('DE.shinyrep.helpers.R')
loadGlobalVars()
load(SHINYREPS_DE_DESEQ, envir=.GlobalEnv)        # the outcome from the DE analysis
resMM <- local({    # load in a local environment to avoid colliding with the analysis w/o MM
    load(SHINYREPS_DE_DESEQ_MM)
    res
})
``` 


## Raw reads QC ##

The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).


#### Read quality

The “read quality” plot summarizes at each read position the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the Phred scores, the better the base calls. The elements of the plot are as follows: the central red line is the median value, the yellow box represents the inter-quartile range (25-75%), the upper and lower whiskers represent the 10% and 90% points, and the blue line represents the mean quality. The background of the graph divides the y-axis into very good quality calls (green), calls of intermediate quality (yellow), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area. Due to binning of quality score in recent sequencing technologies (e.g. Illumina NextSeq), sequencing qualities may be the same on many position and sample profiles may appear very similar.

```{r FastQC_qual_summarized, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}

if (SHINYREPS_FASTQC_SUMMARIZED) {
    fastqc_summary_plot <- DEhelper.Fastqc.custom(web=F,summarizedPlots=T)
    plot(fastqc_summary_plot$p.qual)
    cat("\n", fill = T)
} else {
    fastqc_summary_plot <- DEhelper.Fastqc.custom(web=F,summarizedPlots=F)
}

```

```{r FastQC_qual, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1}

if (SHINYREPS_FASTQC_SUMMARIZED == FALSE) {
    plot(fastqc_summary_plot$p.qual)
    cat("\n", fill = T)
}

```


#### Sequence bias

The "sequence bias" plot shows the proportion of each base (% T, C, A and G) at each position. In a random library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other. But most RNA-seq libraries show sequence imbalance in the first 10-12 read positions due to RT priming biases, which should however look fairly similar in all samples. In case the nucleotide content is summarized in a heatmap for all samples, gray colored fields represent a balanced nucleotide content.


```{r FastQC_nuc_content_summarized, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}

if (SHINYREPS_FASTQC_SUMMARIZED) {
    plot(fastqc_summary_plot$p.content)
    cat("\n", fill = T)
} else {
    fastqc_summary_plot <- DEhelper.Fastqc.custom(web=F,summarizedPlots=F)
}

```

```{r FastQC_nuc_content, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=round((3*fastqc_summary_plot$no.of.samples)/4)+1}

if (SHINYREPS_FASTQC_SUMMARIZED == FALSE) {
    plot(fastqc_summary_plot$p.content)
    cat("\n", fill = T)
}

```


#### GC content

The "GC content" plot shows the GC% distribution of all reads and all samples. Typically, the lines should tightly overlap and look essentially the same in all samples. An unusually shaped distribution could indicate a contaminated library.

```{r FastQC_GC, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}

plot(fastqc_summary_plot$p.gc)
cat("\n", fill = T)

```


```{r cutadapt, echo=F, results='asis', error=F, warning=F, message=F, eval=SHINYREPS_RUN_CUTADAPT=="true"}
cat("## Adapter trimming with Cutadapt\n")
cat("\nThe following plot shows the amount of reads trimmed (in percent) for the selected adapter sequences. The column 'tooshort' gives the percentage of reads removed due to the minimum length threshold after trimming. Additionally to the column 'trimmed' which is for all adapters combined, there are columns for every single adapter giving the percentage of reads trimmed for this adapter.\n\n")

DEhelper.cutadapt()
```


```{r FastQC_qual_summarized_trimmed, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', eval=SHINYREPS_RUN_CUTADAPT=="true"}
cat("### FastQC of trimmed reads QC\n\n")
cat("#### Read quality\n\n")
if (SHINYREPS_FASTQC_SUMMARIZED) {
    fastqc_summary_plot <- DEhelper.Fastqc.custom(web=F,summarizedPlots=T, subdir="trimmed")
    plot(fastqc_summary_plot$p.qual)
    cat("\n", fill = T)
} else {
    fastqc_summary_plot <- DEhelper.Fastqc.custom(web=F,summarizedPlots=F, subdir="trimmed")
}
```

```{r FastQC_qual_trimmed, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=min(170, round((3*fastqc_summary_plot$no.of.samples)/4)+1), eval=SHINYREPS_RUN_CUTADAPT=="true"}

if (SHINYREPS_FASTQC_SUMMARIZED == FALSE) {
    plot(fastqc_summary_plot$p.qual)
    cat("\n", fill = T)
}
```

```{r FastQC_nuc_content_summarized_trimmed, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', fig.height=min(170, ceiling(0.12*fastqc_summary_plot$no.of.samples)+1), eval=SHINYREPS_RUN_CUTADAPT=="true"}
cat("#### Sequence bias of trimmed reads\n")

if (SHINYREPS_FASTQC_SUMMARIZED) {
    plot(fastqc_summary_plot$p.content)
    cat("\n", fill = T)
} 
```

```{r FastQC_nuc_content_trimmed, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=min(170, round((3*fastqc_summary_plot$no.of.samples)/4)+1), eval=SHINYREPS_RUN_CUTADAPT=="true"}

if (SHINYREPS_FASTQC_SUMMARIZED == F) {
    plot(fastqc_summary_plot$p.content)
    cat("\n", fill = T)
}
```

```{r FastQC_GC_trimmed, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center', eval=SHINYREPS_RUN_CUTADAPT=="true"}
cat("#### GC content of trimmed reads\n")
fastqc_summary_plot$p.gc
cat("\n", fill = T)
```


## Read mapping ##

Mapping to the reference genome & transcriptome is performed with STAR (https://github.com/alexdobin/STAR). The program version, genome assembly and software parameters are described in the table at the end of the report.

The mapping statistics below show the number and percentage of: (i) input raw reads, (ii) uniquely mapped reads, (iii) multi-mapped reads aligning equally well to multiple (up to 10) positions in the genome, (iv) reads that align to too many (>10) genome loci are discarded, (v) unmapped reads. For the multimapped reads, one random alignment among the best mapping positions is retained. 

While both unique and multi-mapped reads are included in the generation of browser coverage tracks, only uniquely mapped reads are typically taken into account in the DESeq2 differential expression analysis in order to avoid potential false-positives. However, this standard analysis approach may also exclude some recently duplicated genes or gene-pseudogene pairs, resulting in false-negatives. To investigate the potential expression changes in such genes, a side branch of DESeq2 is run using all mapped reads and the results of both branches are compared.

```{r STAR_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", out.width="100%", fig.height=(fastqc_summary_plot$no.of.samples/5)+2}

star_stats <- DEhelper.STAR()

cat("#### Mapping summary:\n", fill=T)
plot(star_stats$p_perc_count)
cat("\n", fill = T)

cat("#### Mapping statistics:\n", fill=T)
cat(star_stats$stat, sep="\n")
cat("\n", fill = T)

```


## Potential contaminants and rRNA content ##

A competitive mapping to the reference genome and known rRNAs (H. sapiens, M. musculus, C. elegans, D. melanogaster, D. rerio, X. tropicalis) as well as frequently occurring Mycoplasma species (M. arginini, M. fermentans, M. hominis, M. hyorhinis, M. orale, and Acholeplasma laidlawii), PhiX and ERCC spike-in sequences, E. coli and B. taurus (bovine) was performed with FastQScreen (http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/).

In the following plot only relevant contaminants are shown. A contaminant is considered relevant if it consumes `r SHINYREPS_FASTQSCREEN_PERC`% of reads or more in at least one sample. The "one genome" label refers to the amount of reads mapping to that genome only, no matter if they are mapping uniquely or multiple times. The "multiple genomes" label refers to the amount of read mapping to this genome but also to other genomes in the screening regardless of whether they are uniquely or multi-mapping. 

```{r fastqscreen_paragraph, echo=F, results='asis', fig.width=10, error=F, warning=F, message=F}
fastqscreen_plots <- DEhelper.fastqscreen(perc.to.plot=SHINYREPS_FASTQSCREEN_PERC)
```

```{r fastqscreen_paragraph2, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center',fig.height=(round((3*fastqscreen_plots$no.of.genomes)/4)+1)*fastqscreen_plots$no.of.samples/8+1}
  if(is.null(fastqscreen_plots$errortext)) {
    plot(fastqscreen_plots$p.category.wrap)
    } else {
      cat(fastqscreen_plots$errortext)
      }
```


## UCSC genome browser tracks ##

The browser tracks are generated using bamCoverage tool in [deepTools](https://deeptools.readthedocs.io/en/develop/) and they are normalized using the Counts Per Million mapped reads (CPM) method.


To view the tracks in the UCSC Genome Browser, navigate to the [UCSC Genome Browser Track Hub page](https://genome-euro.ucsc.edu/cgi-bin/hgHubConnect) and enter the following line into the `URL` field:

```{r Trackhub, echo=F, results='asis', error=F, warning=F, message=F}
cat(DEhelper.Trackhub(), sep="\n")
```


## Insert size (for paired-end sequencing) ##

The insert size is defined as the distance between the left-most mapping position of a read and the right-most mapping position of its mate. For DNA, this exactly represents the fragment length. However, in case of RNA, the estimated insert size can be a lot larger than the true insert size / fragment length, when e.g. the read pair overlaps introns or skipped exons. Thus, the insert size calculation can only be seen as a rough estimate of the true insert size / fragment length when looking at RNAseq data.

```{r echo=F, error=F, fig.width=15, message=FALSE, warning=FALSE, results='asis'}

if (!file.exists(SHINYREPS_INSERTSIZE)) {
    cat("Insert size data not available.")
    } else {
    cat(DEhelper.insertsize(), sep="\n")
    p <- DEhelper.insertsize.plot()
    DEhelper.subchunkify(p, 
                         fig_height = dim(p)[1] * 15/dim(p)[2],
                         fig_width = 15)
    }

```


## Library complexity assessment ##

Measuring the fraction of duplicated reads is commonly used to control for excessive PCR over-amplification during library preparation and/or for suboptimal amount of input material. However, in RNA-seq duplicated reads may also arise naturally due to the highly expressed (e.g. housekeeping) genes, which makes the overall read duplication rate useless. The IMB-developed tool dupRadar (http://bioconductor.org/packages/release/bioc/html/dupRadar.html) gives an informative insight into the duplication problem by graphically relating the gene expression levels and the observed duplication rates. Thus, problematic samples can be easily identified by visual inspection. Good quality samples with high library complexity show a sigmoidal curve with low duplication rates of weakly expressed genes (below 100 reads/kbp).

```{r Dupradar_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(DEhelper.dupRadar(web=F), sep="\n")
```


## Mapped reads annotation ##

The mapped reads were assigned to annotated features (gene exons) using the Subread tool featureCounts (http://bioinf.wehi.edu.au/featureCounts). The unassigned reads are due to ambiguity (overlap with multiple features), non-unique mapping (multimap) or because no overlap is found with known exonic features (e.g. intronic reads derived from unprocessed transcripts). Typically, mRNA-seq with polyA-selection results in more exon-assigned reads compared to total RNA-seq with rRNA-depletion. 

```{r Subread_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(DEhelper.Subread(), sep="\n")
```


## RNA class representation ##

The following plot shows the fraction of reads assigned to various RNA classes. The plot can help in determining if the sample prep protocol worked consistently well for all samples and may reveal issues with e.g. rRNA contamination.

```{r RNAtypes_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", out.width="100%"}
DEhelper.RNAtypes()
```


## Gene body coverage ##

The plots below show meta-gene profiles of 5' to 3' read coverage and can reveal biases due to RNA degradation or specific library prep protocol. PolyA-selection protocols are particularly prone of producing 3'-coverage bias due to even mild RNA degradation. Ideally, all samples which are being compared have similar profiles.

```{r GeneBodyCoverage_paragraph, echo=F, error=F, warning=F, message=F, results='asis', fig.align="center"}
plot_list <- DEhelper.geneBodyCov2(web=F)
cat("#### Gene body coverage per sample: plotly")
cat("\n\n", fill=T)
plot_list$plotly %>%
    highlight(dynamic=T, selectize=T, persistent=T) 
cat("\n\n")

if("p_per_sample" %in% names(plot_list)){
   cat("#### Gene body coverage per sample: \n", fill=T)
   plot(plot_list$p_per_sample)
   cat("\n", fill=T)
}

if("num_groups" %in% names(plot_list)) {
   no.of.groups <- plot_list$num_groups
} else {
   no.of.groups <- 1
}

if("num_replicates" %in% names(plot_list)) {
   no.of.replicates <- plot_list$num_replicates
} else {
   no.of.replicates <- 1
}
```

```{r GeneBodyCoverage_splitByGroup, echo=F, error=F, warning=F, message=F, results='asis', fig.align="center", out.width="100%", fig.height=round((no.of.groups+1)/3)*2.5+0.5+ifelse(no.of.groups<3,1,0)}
if("p_per_sample_splitByGroup" %in% names(plot_list)){
   cat("#### Gene body coverage per sample split by group: \n", fill=T)
   plot(plot_list$p_per_sample_splitByGroup)
   cat("\n", fill=T)
}
if("p_per_replicate_splitByGroup" %in% names(plot_list)){
  cat("#### Gene body coverage per sample split by group, colored by replicate: \n", fill=T)
  plot(plot_list$p_per_replicate_splitByGroup)
  cat("\n", fill=T)
}
```

```{r GeneBodyCoverage_splitByReplicate, echo=F, error=F, warning=F, message=F, results='asis', fig.align="center", out.width="100%", fig.height=round((no.of.replicates+1)/3)*2.5+0.5+ifelse(no.of.replicates<3,1,0)}
if("p_per_group_splitByReplicate" %in% names(plot_list)){
  cat("#### Gene body coverage per sample split by replicate, colored by group: \n", fill=T)
  plot(plot_list$p_per_group_splitByReplicate)
  cat("\n", fill=T)
}
```


## Strand specificity ##

The table below shows the fraction of reads mapped in sense or antisense to gene exons - should be around 0.5 for non-stranded library prep protocols and close to 0 or 1 for strand-specific RNA-seq protocols. The strandedness is calculated as the percentage of unambiguous reads that are on the strand expected by the library preparation protocol.

```{r echo=F, results='asis', error=F, warning=F, message=F}
cat(DEhelper.strandspecificity(), sep="\n")
```


## Distribution of reads ##

The following plot depicts the fraction of reads mapped to exons, introns and intergenic regions. For mRNA-seq in well-annotated genomes the fraction of exonic reads is typically quite high, whereas for total RNA-seq intronic and intergenic reads may dominate.

```{r echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", out.width="100%", fig.height=(fastqc_summary_plot$no.of.samples/5)+1.5}
DEhelper.Qualimap()        # sample Qualimap 
```

Please note that reads overlapping an exon as well as an intron/intergenic region are counted differently by the tools used in this section (QualiMap) and the tool used for assignment of reads in section “Mapped reads annotation” (featureCounts). In the latter, all reads overlapping an exon by at least one base pair are considered exonic, while in the current section only reads fully located in exons are considered exonic and all reads overlapping an intron or intergenic region by at least one bp are considered intronic/intergenic. Thus, numbers do not fully agree between those two sections.


## Sample relatedness ##

The plot below depicts sample similarity based on priniciple component analysis (PCA) with the 500 genes showing the highest variance. Overall, replicates are expected to show relatively small distances between them, while sample groups often show higher distances between them, depending on the magnitude of expression changes. If the samples cluster other than by group, there might be sample prep batch effects or some other unaccounted biological/technical factors in the experiment. 

```{r MDSplot_paragraph, echo=F, error=F, warning=F, message=F, fig.align="center", out.width="100%"}
DEhelper.DESeq2.MDS()        # sample MDS 
```

A heatmap clustogram of the Euclidean distances between the samples gives an intuitive measure of sample similarity:

```{r Heatmap2_paragraph, echo=F, error=F, warning=F, message=F, fig.align="center"}
DEhelper.DESeq2.heatmap(type="distance")         # heatmap of sample to sample distances
```

Additionally, hierarchical clustering of the samples based on the most variable genes reveals replicate consistency and can spot batches of samples related by other characteristics rather than condition:

```{r Heatmap1_paragraph_sd, echo=F, error=F, warning=F, message=F, fig.align="center"}
DEhelper.DESeq2.heatmap(type="cluster_sd")   # heatmap of top variant 'n' genes of the rlog transformed counts (normalized and stabilized)
```

Finally, a hierarchical clustering of the samples based on the genes with the highest average expression level:

```{r Heatmap1_paragraph_mean, echo=F, error=F, warning=F, message=F, fig.align="center"}
DEhelper.DESeq2.heatmap(type="cluster_mean")    # heatmap of top variant 'n' genes of the rlog transformed counts (normalized and stabilized)
```


## Differential gene expression analysis ##

Pairwise differential expression analysis on the gene level between the experimental groups is performed with the popular R/Bioconductor package DESeq2 (https://bioconductor.org/packages/release/bioc/html/DESeq2.html). Default DESeq2 analysis includes an independent filtering procedure for lowly expressed genes which is meant to ameliorate the multiple testing problem thereby maximising the statistical power to detect differentially expressed genes (DEGs). In this standard analysis only uniquely mapped reads are taken into account to avoid "cross-talk" between different genes. 

Note that by default a standard DESeq2 analysis uses only a statistical (FDR) filter and is thus testing for evidence that the expression changes between two sample groups is different from zero. Performing post-hoc filtering of the results by effect size (fold-change) is not recommended by the software developers and should be avoided. Such a "Volcano plot" gene filtering approach may result in inflated false discovery rates as reported in https://doi.org/10.1093/bib/bbab053. If desirable, the differential expression analysis can incorporate an additional fold-change (FC) threshold directly in the DESeq2 statistical testing model, which results in more conservative but accurate DEG estimates.

For the analyses below, the FDR filter was set to `r SHINYREPS_DE_DESEQ_FDR` and the effect size (FC) filter was set to `r SHINYREPS_DE_DESEQ_FC` (note: FC=1 means no fold-change filter was used).

```{r DifferentialExpression_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center"}
for(i in 1:nrow(conts)) {

    # subsection
    cont.name <- conts$contrast.name[i]
    cat("### ", cont.name, fill=T)
    cat("", fill=T)

    # pairwise PCA plot 
    cat("##### PCA plot", fill=T)
    cat("Priniciple component analysis (PCA) plot using only the samples of the two analysed groups:", "\n", "\n")
    DEhelper.DESeq2.pairwisePCA(i)
    cat("\n", fill=T)

    # sample to sample distance
    cat("##### Sample to sample distance (diagonal set to NA)", fill=T)
    DEhelper.DESeq2.heatmap(i, dds=pairwise.dds,  logTransform=T, type="distance")
    cat("\n", fill=T)

    # most variable genes
    cat("##### Most variable genes", fill=T)
    DEhelper.DESeq2.heatmap(i, dds=pairwise.dds,  logTransform=T, type="cluster_sd")
    cat("\n", fill=T)

    # MA plot
    cat("##### MA plot", fill=T)
    fdr <- as.numeric(SHINYREPS_DE_DESEQ_FDR)
    FC  <- as.numeric(SHINYREPS_DE_DESEQ_FC)
    genes <- DEhelper.DESeq2.DEgenes(i)
    #filtering for tested genes
    genes <- genes[!is.na(genes$padj), ]
    cat("MA plot showing expression fold change on log2 scale vs. mean expression level of all annotated genes. There are", 
        sum((genes$padj < fdr) & (abs(genes$log2FoldChange) > log2(FC)), na.rm=T), "genes with a minimal FC of", FC, "highlighted in the MA plot with corrected pvalue (Benjamini and Hochberg, FDR) <", fdr, ".", "\n")
    #we do not need an additional filter for the FC here
    #since only genes with a minimal fold change of FC are
    #significant for the significance level chosen in the analysis
    DEhelper.DESeq2.MAplot(i, fdr)    # should be fine at taking the value from the UI slider
    cat("\n", fill=T)

    # Volcano plot
    cat("##### Volcano plot", fill=T)    
    cat("Volcano plot of all DESeq2-tested genes depicting their expression fold change vs. statistical significance. Differentially expressed genes are highlighted in red and the top 25 are labelled.", "\n", "\n")
    DEhelper.DESeq2.VolcanoPlot(i, fdr, web=F)
    cat("\n", fill=T)

    # and DE table
    cat("##### Top hits", fill=T)
    cat("", fill=T)
    cat("The table below shows the top 25 DEGs with their mean normalised counts (baseMean), expression change (log2FoldChange) and statistical significance (padj = FDR-adjusted p-value). The complete DESeq2 results table including FPKM-normalised expression data for all genes in each sample is available in CSV and Excel format in the 'DE_DESeq2' results folder.", fill=T)
    cat("", fill=T)
    genes$baseMean       <- round(genes$baseMean, 2)
    genes$log2FoldChange <- round(genes$log2FoldChange, 2)
    genes$padj           <- format(genes$padj, digits=3, scientific=TRUE)
    cols <- c("gene_name", "baseMean", "log2FoldChange", "padj")
    cat(kable(genes[1:25, cols], row.names=T, align="r", format="markdown"), sep="\n")    # show only the top 25 genes
    cat("", fill=T)

}
```


## GO and Pathway enrichment analysis ##

The R/Bioconductor package clusterProfiler (https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) was used to find enriched GO terms and pathways in the lists of significantly up- and down-regulated genes from the pairwise differential expression analyses. The resulting enrichment tables and plots can be found in the 'GO_Analysis' results folder. Shown below are several complementary visualisations aimed to help understand the underlying biological changes and the mutual relationships between the enriched GO terms and pathways.

```{r Enrichment_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center"}

for(i in 1:nrow(conts)) {

    # subsection
    cont.name <- gsub("(.+)=(.+)","\\1",conts[i,1])
    cat("### ", cont.name, fill=T)
    cat("", fill=T)

    # GO Enrichment Plots
    # up-regulated genes
    cat("##### GO terms enriched in the list of up-regulated genes (Biological Process, BP)", fill=T)

    GO_bar_up     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_barplot_up_genes.png")
    GO_network_up <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_network_up_genes.png")
    GO_treemap_up <- paste0(SHINYREPS_GO_ENRICHMENT, "/", conts$contrast.name[i], "_GO_treemap_up_genes.png")
    GO_scatter_up <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_scatterplot_up_genes.png")
    GO_dot_up     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_reduced_GO_dotplot_up_genes.png")
    GO_cnet_up    <- paste0(SHINYREPS_GO_ENRICHMENT, "/", conts$contrast.name[i], "_reduced_GO_cnetplot_up_genes.png")

    # if there are GO enrichment results (the barplot was created)
    if(file.exists(GO_bar_up)) {

        # GO enriched terms in a barplot
        cat("Top enriched Gene Ontology BP terms for the up-regulated genes:", "\n", fill=T)
        cat("![***](", GO_bar_up, ")\n", fill=T)

        # GO enriched terms in a network
        cat("Network plot of the top enriched Gene Ontology BP terms for the up-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", GO_network_up, ")\n", fill=T)

        # GO enriched terms in a treemap
        cat("Clustering and reduction of the enriched Gene Ontology BP terms for the up-regulated genes (area size correlates with significance):", "\n", fill=T)
        cat("![***](", GO_treemap_up, ")\n", fill=T)

        # reduced GO enriched terms in a dotplot
        cat("Top enriched Gene Ontology BP term clusters (reduced GO terms) for the up-regulated genes:", "\n", fill=T)
        cat("![***](", GO_dot_up, ")\n", fill=T)

        # GO enriched terms in a cnetplot
        cat("Top enriched Gene Ontology BP term clusters with associated genes for the up-regulated genes:", "\n", fill=T)
        cat("![***](", GO_cnet_up, ")\n", fill=T)

    } else {
        cat("no GO enrichment results for the up-regulated genes available\n", fill=T)
    }

     # Pathway Enrichment Plots
     # up-regulated genes
     cat("##### Reactome and KEGG pathways enriched in the list of up-regulated genes", fill=T)

     Reactome_Pathway_bar_up     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i,1]),"_Reactome_barplot_up_genes.png")
     Reactome_network_up         <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_Reactome_network_up_genes.png")

     # if there are Reactome Pathway enrichment results (the barplot was created)
     if(file.exists(Reactome_Pathway_bar_up)) {

        # Reactome Pathways enriched in a barplot
        cat("Top enriched Reactome pathways for the up-regulated genes:", "\n", fill=T)
        cat("![***](", Reactome_Pathway_bar_up, ")\n", fill=T)

        # Reactome enriched terms in a network
        cat("Network plot of the top enriched Reactome pathways for the up-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", Reactome_network_up, ")\n", fill=T)

     } else {
         cat("no Reactome Pathway enrichment results for the up-regulated genes available\n", fill=T)
    }

    KEGG_Pathway_bar_up     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i,1]),"_KEGG_barplot_up_genes.png")
    KEGG_network_up         <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_KEGG_network_up_genes.png")

    # if there are KEGG Pathway enrichment results (the barplot was created)
    if(file.exists(KEGG_Pathway_bar_up)) {

        # KEGG Pathways enriched in a barplot
        cat("Top enriched KEGG pathways for the up-regulated genes:", "\n", fill=T)
        cat("![***](", KEGG_Pathway_bar_up, ")\n", fill=T)

        # KEGG enriched terms in a network
        cat("Network plot of the top enriched KEGG pathways for the up-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", KEGG_network_up, ")\n", fill=T)

     } else {
         cat("no KEGG Pathway enrichment results for the up-regulated genes available\n", fill=T)
    }

    # GO Enrichment Plots
    # down-regulated genes
    cat("##### GO terms enriched in the list of down-regulated genes (Biological Process, BP)", fill=T)

    GO_bar_down     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_barplot_down_genes.png")
    GO_network_down <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_network_down_genes.png")
    GO_treemap_down <- paste0(SHINYREPS_GO_ENRICHMENT, "/", conts$contrast.name[i], "_GO_treemap_down_genes.png")
    GO_scatter_down <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_GO_scatterplot_down_genes.png")
    GO_dot_down     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_reduced_GO_dotplot_down_genes.png")
    GO_cnet_down    <- paste0(SHINYREPS_GO_ENRICHMENT, "/", conts$contrast.name[i], "_reduced_GO_cnetplot_down_genes.png")

    # if there are GO enrichment results (the barplot was created)
    if(file.exists(GO_bar_down)) {

        # GO enriched terms in a barplot
        cat("Top enriched Gene Ontology BP terms for the down-regulated genes:", "\n", fill=T)
        cat("![***](", GO_bar_down, ")\n", fill=T)

        # GO enriched terms in a network
        cat("Network plot of the top enriched Gene Ontology BP terms for the down-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", GO_network_down, ")\n", fill=T)

        # GO enriched terms in a treemap
        cat("Clustering and reduction of the enriched Gene Ontology BP terms for the down-regulated genes (area size correlates with significance):", "\n", fill=T)
        cat("![***](", GO_treemap_down, ")\n", fill=T)

        # reduced GO enriched terms in a dotplot
        cat("Top enriched Gene Ontology BP term clusters (reduced GO terms) for the down-regulated genes:", "\n", fill=T)
        cat("![***](", GO_dot_down, ")\n", fill=T)

        # GO enriched terms in a cnetplot
        cat("Top enriched Gene Ontology BP term clusters with associated genes for the down-regulated genes:", "\n", fill=T)
        cat("![***](", GO_cnet_down, ")\n", fill=T)

    } else {
        cat("no GO enrichment results for the down-regulated genes available\n", fill=T)
    }

     # Pathway Enrichment Plots
     # down-regulated genes
     cat("##### Reactome and KEGG pathways enriched in the list of down-regulated genes", fill=T)

     Reactome_Pathway_bar_down     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i,1]),"_Reactome_barplot_down_genes.png")
     Reactome_network_down         <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_Reactome_network_down_genes.png")

     # if there are Reactome Pathway enrichment results (the barplot was created)
     if(file.exists(Reactome_Pathway_bar_down)) {

        # Reactome Pathways enriched in a barplot
        cat("Top enriched Reactome pathways for the down-regulated genes:", "\n", fill=T)
        cat("![***](", Reactome_Pathway_bar_down, ")\n", fill=T)

        # Reactome enriched terms in a network
        cat("Network plot of the top enriched Reactome pathways for the down-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", Reactome_network_down, ")\n", fill=T)

     } else {
         cat("no Reactome Pathway enrichment results for the down-regulated genes available\n", fill=T)
    }

    KEGG_Pathway_bar_down     <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i,1]),"_KEGG_barplot_down_genes.png")
    KEGG_network_down         <- paste0(SHINYREPS_GO_ENRICHMENT, "/", gsub("=.*", "", conts[i, 1]), "_KEGG_network_down_genes.png")

    # if there are KEGG Pathway enrichment results (the barplot was created)
    if(file.exists(KEGG_Pathway_bar_down)) {

        # KEGG Pathways enriched in a barplot
        cat("Top enriched KEGG pathways for the down-regulated genes:", "\n", fill=T)
        cat("![***](", KEGG_Pathway_bar_down, ")\n", fill=T)

        # KEGG enriched terms in a network
        cat("Network plot of the top enriched KEGG pathways for the down-regulated genes (connecting lines represent shared genes):", "\n", fill=T)
        cat("![***](", KEGG_network_down, ")\n", fill=T)


    } else {
        cat("no KEGG Pathway enrichment results for the down-regulated genes available\n", fill=T)
    }

}
```


## Chromosome over-representation analysis ##

In some experiments it is possible that there is local effect in gene expression due to for instance genetic linkage. One hint that this might be happening is if there are more affected genes in a chromosome than expected by chance. If this happens, further data exploration might be warranted. 

The tables contain a summary of the likelihood that there are more DE genes in a particular chromosome than expected by chance. The p-value is calculated with Fisher's exact test and corrected for multiple testing (FDR). Only chromosomes which show a significant over-representation (FDR < 0.1) are presented in this table. This test also gives an odds ratio which represents the strength of association. The odds ratio compares the number of DE genes found on chromosome A with its expected number. If an odds ratio is equal to or less than 1, there is no association between the two lists. If the odds ratio is much larger than 1, then the association is strong. The class also calculates the Jaccard index which measures the similarity between two lists of genes. Here, it is comparing the list of DE genes on a chromosome A with the list of genes, which are either DE and/or on chromosome A. The Jaccard index varies between 0 and 1, with 0 meaning there is no similarity between the two and 1 meaning the two are identical.


```{r ChromosomeEnrichment_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center"}
for(i in 1:nrow(conts)) {

    # subsection
    cont.name <- conts$contrast.name[i]
    cat("### ", cont.name, fill=T)    
    cat("", fill=T)
    
    # and DE table
    table_fisher <- DEhelper.DESeq2.ChrOverrepresentation(i, fdr)

    if (nrow(table_fisher) > 0) {
        cat(kable(table_fisher, row.names=T, align="r", format="markdown"), sep="\n")
        cat("", fill=T)
    } else {
        cat("No chromosome over-representation found.", fill=T)
        cat("", fill=T)
    }
}
```


## Differential splicing analysis ##

Differential splicing analysis is performed with the rMATS software suite (http://rnaseq-mats.sourceforge.net/), which uses the RNA-Seq data to detect differential alternative splicing events. The statistical model of rMATS calculates the probability that the difference in the isoform ratio of a gene between two conditions exceeds an user-defined threshold (Default: 0.01%). It is important to note that, rMATS is designed to analyze only the basic types of alternative splicing events involving two isoforms from an alternatively spliced region. These types include exon skipping (SE), alternative 5' splice sites (A5SS), alternative 3' splice sites (A3SS), mutually exclusive exons (MXE), and retained introns (RI). 

Further, the maser package is used to analyze the splicing events. Low coverage splicing junctions are commonly found in RNA-seq data and lead to low confidence differential percent spliced-in PSI levels. Hence, low coverage events with coverage < `r SHINYREPS_MASER_MINCOV` reads are removed. Further, statistically significant events are selected using a FDR cutoff of `r SHINYREPS_MASER_FDR` and minimum PSI change of `r as.numeric(SHINYREPS_MASER_DPSI)*100`% (deltaPSI = `r SHINYREPS_MASER_DPSI`).

For each pairwise analysis, a distribution of the proportion of splicing events are shown as a summary. Further, for each of the aforementioned alternative splicing events (A3SS, A5SS, SE, RI, MXE) the following global analysis plots are shown, which can also be found in the 'rMATS' results folder: 
  
  * a table of genes with significant splicing events, their exon inclusion levels (PSI_1, PSI_2, deltaPSI) and exon localization,
  * a volcano plot of the splicing events with only the significant events colored (blue, red).

In addition, only for the gene with the *top* significant splicing event, we show a customized genomic track visualization of the transcripts affected by the respective splicing event. More information on interpreting that plot is provided below. 

```{r Diff_Splicing, echo=F, results='asis', error=F, warning=F, message=F}
# load contrasts
conts <- read.delim(file.path(SHINYREPS_PROJECT, "contrasts.txt"), header=T,comment.char="#")

# extract only unique contrasts
# duplicates may occur, if user tries different formula for same contrasts
uconts <- unique(conts[,c("contrast.name","contrast")])
for(i in 1:nrow(uconts)) {
	# subsection
	cont.name <- uconts$contrast.name[i]
	if(!dir.exists(file.path(SHINYREPS_RES, "rMATS", paste0(cont.name,"_rMATS")))) {next()}

	cat("\n\n", fill=T)
	cat("### ", cont.name, fill=T)
	cat("\n\n", fill=T)
	if(file.exists(file.path(SHINYREPS_RES, "rMATS", paste0(cont.name,"_rMATS"),"maser_processed.RData"))) {
		load(file.path(SHINYREPS_RES, "rMATS", paste0(cont.name,"_rMATS"),"maser_processed.RData"))
		DEhelper.rmats()
	} else {
		cat("No splicing events were found")
		next()
	}
}
```


## Additional DE analysis including multimapped reads ##

For many genes in the human, mouse and other genomes, RNA-seq with short reads cannot accurately measure expression because recently duplicated gene families and gene-pseudogene pairs have very similar transcript sequences, which results in mostly ambiguous or multimapped reads. 

There are several approaches how to deal with multimapped reads, here they are randomly assigned to one of the possible source genes followed by DESeq2 statistical assessment. While this alternative differential expression analysis may reveal additional interesting hits, it may also produce false-positive findings due to potential cross-talk between highly similar genes, i.e. if a gene X changes expression, its pseudogene Y or closely-related paralog Z could be non-specifically affected as well. Therefore, the standard DESeq2 analysis using only uniquely mapped reads is considered more reliable for the majority of genes and should have priority.

The Venn diagrams below compare the number of differentially expressed genes detected with and without inclusion of the ambiguous multimapped (MM) reads.

```{r DifferentialExpressionMM_paragraph, echo=F, results='asis', error=F, warning=F, message=F, fig.align="center", fig.width=3, fig.height=3}
for(i in 1:nrow(conts)) {

    fdr <- .01
    
    # subsection
    cont.name <- conts$contrast.name[i]
    cat("### ", cont.name, fill=T)
    cat("", fill=T)
    
    # Venn comparing DE genes w/ w/o MM
    padj   <- ifelse(is.na(res  [[i]]$padj), 1, res  [[i]]$padj)
    padjMM <- ifelse(is.na(resMM[[i]]$padj), 1, resMM[[i]]$padj)
    if (sum(padj < fdr) > 0 && sum(padjMM < fdr) > 0) { 
        futile.logger::flog.threshold(futile.logger::ERROR, name = "VennDiagramLogger")  # to get rid of Venn plot logs
        img <- venn.diagram(list(with_MM  =rownames(resMM[[i]])[padjMM < fdr],
                                 without_MM=rownames(res  [[i]])[padj   < fdr]),
                                 fill=c("red", "blue"), filename=NULL, alpha=0.3, lwd=0,
                                 margin=0.1, height=1200, ext.pos=180, ext.length = 0.8, cat.dist=0.1)
        grid.newpage()
        grid.draw(img)
    }
    
    # the top genes in disagreement
    cat("The table below shows the top genes which are differentially expressed (FDR <", fdr, ") in only one analysis.", fill=T)
    cat("", fill=T)
    disagree   <- setdiff(rownames(res  [[i]])[padj   < fdr], rownames(resMM[[i]])[padjMM < fdr])
    disagreeMM <- setdiff(rownames(resMM[[i]])[padjMM < fdr], rownames(res  [[i]])[padj   < fdr])
    tables <- lapply(list(disagree, disagreeMM), function(genes) {
        out <- data.frame(name       =resMM[[i]][genes, "gene_name"],
                          baseMean   =round (res[[i]][genes, "baseMean"], 2),
                          log2FC     =round (res[[i]][genes, "log2FoldChange"], 2),
                          padj       =format(res[[i]][genes, "padj"], digits=3, scientific=TRUE),
                          baseMean_MM=round (resMM[[i]][genes, "baseMean"], 2),
                          log2FC_MM  =round (resMM[[i]][genes, "log2FoldChange"], 2),
                          padj_MM    =format(resMM[[i]][genes, "padj"], digits=3, scientific=TRUE))
        rownames(out) <- genes
        o <- order(abs(out$log2FC - out$log2FC_MM), decreasing=T)
        out[o[1:min(nrow(out), 25)], ]
    })

    # output result
    cat("With multimappers only:", fill=T)
    cat("", fill=T)
    cat(kable(tables[[2]], row.names=T, align="r",format="markdown"), sep="\n")    # show max 25 genes
    cat("", fill=T)
    cat("Without multimappers only:", fill=T)
    cat("", fill=T)
    cat(kable(tables[[1]], row.names=T, align="r",format="markdown"), sep="\n")    # show max 25 genes
    cat("", fill=T)
}
```

## Used tools and versions for this analysis ##

Read mapping was performed with STAR using the following parameters:

```{r STAR_parameters_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(DEhelper.STARparms(), sep="\n")
``` 

The following tools were also used for data processing:

```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

</div>

