---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: yes
    css: styles.css
---

<div class="contentbox">

```{r LoadingGloabalVariables, echo=F,result='hide',error=F,warning=F,message=F}
require(rmarkdown)
require(knitr)
require(Cairo)                                                    
knitr::opts_chunk[['set']](dev='CairoPNG')
source('variant.shinyrep.helpers.R')
loadGlobalVars()
``` 

## Raw reads QC ##

### Raw reads qualities, sequence bias and duplication

The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

1. The "Read qualities" heatmap plot summarizes the mean quality score at each position.
The color on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (yellow), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area.

2. The "Sequence bias" heatmap plot shows the proportion of each base (% G, A, T and C) at each position. In a random library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other. But most RNA-seq libraries show sequence imbalance in the first 10-12 read positions due to RT priming biases, which should however look fairly similar in all samples.

3. The "GC Content" plot shows the GC% distribution of all reads and the ideal normal distribution based on the data (black line). Typically, the lines should tightly overlap and look essentially the same in all samples. An unusually shaped distribution could indicate a contaminated library.

```{r FastQC_paragraph, echo=F,results='asis',error=F,warning=F,message=F,fig.align='center'}
VARhelper.ngsReports.Fastqc()
```

### Mapping
The mapping statistics show the total number of reads delivered to the aligner, the number of mapped reads and reads not mapping at all to the ref aligning to the wrong reference genome.

```{r BWA_ReadStats_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(VARhelper.BWA(),sep="\n")
```

### Variant calling
Variant calling can be performed using GATK UnifiedGenotyper and GATK HaplotypeCaller.
Both tools include default read filters that exclude reads from variant calling.

#### GATK UnifiedGenotyper filtered read statistics
```{r GATKug_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(VARhelper.GATKug(),sep="\n")
```

#### GATK HaplotypeCaller filtered read statistics

```{r GATKhc_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(VARhelper.GATKhc(),sep="\n")
```

GATK Haplotyper annotates variants with dbSNP entries, if the database file is provided. Novel variants, which are not found in dbSNP usually tend to be heterozygous variants in low coverage regions.
These regions are prone to lead to false positive calls and should be taken with a grain of salt.

```{r CoveragePlot, echo=F,results='asis',error=F,warning=F,message=F}
no.feedback <- VARhelper.CoveragePlot()
```

#### GATK VariantEval variant statistics

Variant statistics show the basic properties of your variants as triplets for "all", "known" ,"novel" variants in comparison to dbSNP.
The calculated ratios should be similar in "all", "known" and "novel" subsets. Highly deviating values in the "novel" subset indicate a high false positive rate.
Due to selective pressure transitions are more likely to be retained in the genome than transversions.
For (human) Exome sequencing a Ti/Tv ration of 2.8-3 would be good, while whole genome sequencing might result in values 2-2.1. A completely random Ti/Tv ratio would be 0.5.
Sample names end with .HC (HaplotypeCaller) and .UG (UnifiedGenotyper).

[one source of many](http://www.cureffi.org/2012/10/17/descriptive-statistics-and-quality-control-on-variants/)

```{r GATK_VariantEval_paragraph, echo=F,results='asis',error=F,warning=F,message=F}
cat(VARhelper.GATKvarianteval(),sep="\n")
```

## Used tools and versions for this analysis ##

```{r ToolVersions_paragraph, echo=F,results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

</div>
