---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: yes
    css: styles.css
---

<div class="contentbox">

```{r Initialise, echo=F, result='hide', error=F, warning=F, message=F}
options(stringsAsFactors=FALSE)
source('smallRNA.shinyrep.helpers.R')
loadGlobalVars()
``` 

## Quality Control smallRNAseq ##

### Raw reads QC
The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

1. The "Read qualities" Box-Whisker plots show the range of quality values across all base positions:
    (i) The central red line is the median value,
    (ii) The yellow box represents the inter-quartile range (25-75%),
    (iii) The upper and lower whiskers represent the 10% and 90% points,
    (iv) The blue line represents the mean quality.
The y-axis on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area.

2. The "Sequence bias" plots show the proportion of each base (% G, A, T and C) at each position. In a random library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other.


```{r FastQC_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.Fastqc(), sep="\n")
```

### Adapter trimming

Adapter trimming was done using cutadapt (http://journal.embnet.org/index.php/embnetjournal/article/view/200). The minimal overlap of read and adapter was set to `r SHINYREPS_MINADAPTEROVERLAP`. Reads shorter than `r SHINYREPS_MINREADLENGTH` bp (insert + universal molecular identifiers (UMIs)) after adapter trimming were discarded.

```{r cutadapt_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.cutadapt(), sep="\n")
```


### Quality filtering

After trimming off adapters at the 3â€™ end of reads, low quality reads were discarded, i.e. all reads
with any nucleotide having a quality score below 20 were removed.

```{r qualityfilter_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.qualityfilter(), sep="\n")
```


### Duplicate removal

Duplicates are then removed from the remaining high-quality reads. Here, identical reads including
identical UMIs (unique molecular identifiers) are considered duplicates. For each group of duplicates, one copy was kept for subsequent analysis.

```{r dedup_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.dedup(), sep="\n")
```


### Summary of raw data filtering

A summary of all trimming and filtering until removal of duplicates is shown below.

```{r rawfiltersummary_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.rawfiltersummary(), sep="\n")
```



### QC of filtered raw reads

The first two columns show "Read qualities" and "Sequence bias" as described before. The third column "Length distribution" shows
the length distribution of remaining reads after trimming of UMIs on both ends.

```{r filteredFastQC_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.filteredFastqc(), sep="\n")
```



## Read mapping

### Mapping to the genome

Mapping to the reference genome & transcriptome is performed with Bowtie (http://bowtie-bio.sourceforge.net). The program version, genome assembly and software parameters are described in the table at the end of the report.

The mapping statistics below show the number and percentage of (i) input raw reads, (ii) uniquely mapped reads, (iii) reads mapped to multiple positions in the genome (multimapped), and (iv) reads not mapped to the reference genome. For the multimapped reads, one random alignment among the best mapping positions is chosen.

```{r bowtie_table_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.bowtie.table(), sep="\n")
``` 

```{r bowtie_plot_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.bowtie.plot(), sep="\n")
```


### Mapping to rRNAs and the genome

Using FastQScreen (http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/), the amount of reads also mapping to rRNAs was determined.
To that aim, reads were mapped to both; the genome and all known rRNAs. Results are shown below.

```{r fastqscreen_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.fastqscreen(), sep="\n")
```



### Mapped reads annotation
The mapped reads were assigned to annotated features using the Subread tool featureCounts (http://bioinf.wehi.edu.au/featureCounts). The unassigned reads are due to ambiguity (overlap with multiple features) or because no overlap is found with known features (e.g. intronic reads derived from unprocessed transcripts).  

```{r subread_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.subread(), sep="\n")
```

### RNA class representation
The following plot shows the fraction of reads assigned to various RNA classes. All RNA classes occurring with more than `r as.numeric(SHINYREPS_RNATYPES_CUTOFF)*100`% are plotted individually, all others are combined to "other". These plots help in determining if the sample prep protocol worked well and may reveal issues with rRNA contamination. 

```{r RNAtypes_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
smallRNAhelper.RNAtypes()
```



## Used tools and versions for this analysis

Mapping was performed with Bowtie using the following parameters:

```{r bowtie_parameters_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(smallRNAhelper.bowtie.params(), sep="\n")
``` 

The following tools and versions were used for the analysis:

```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

</div>

