---
title: "SHINYREPS_PROJECT"
output:
  html_document:
    toc: true
    toc_float: true
    css: styles.css
---

<div class="contentbox">


# Description

Enter project description here


# Processing

Enter data processing here


```{r setup, echo=F, result='hide', error=F, warning=F, message=F}

# source helper functions
source("sc.shinyrep.helpers.R")

# load required packages
attach_packages(c("rmarkdown", "knitr", "Cairo", "gplots", "ggplot2", "plotly", "RColorBrewer", "ggrepel", "parallel",
                  "grid", "gridExtra", "rtracklayer", "scater", "scran", "ggbeeswarm", "dplyr", "tidyr", "forcats",
                  "scales", "kableExtra", "reshape2", "M3Drop", "vipor", "corrplot", "limma", "pheatmap",
                  "VennDiagram", "edgeR", "DESeq2", "GeneOverlap", "viridis", "Biobase", "batchelor", "SC3", "pkgmaker",
                  "igraph", "pheatmap", "cluster", "dendextend", "dynamicTreeCut", "shinydashboard"))

# load global variables
loadGlobalVars(f="shinyReports.txt")

# create folder for output files
if (!file.exists(file.path("report_files"))) {dir.create(file.path("report_files"))}

# set options
options(stringsAsFactors=FALSE)
CORES <- 2
pal   <- brewer.pal(9, "Set1")
pal_rb <- colorRampPalette(c(pal[1], "white", pal[2]))(20)
pal_y  <- colorRampPalette(c("black", "yellow"))(100)

knitr::opts_chunk$set(cache=F,
                      echo=F,
                      warning=F,
                      message=F,
                      dev='CairoPNG')

theme_set(theme_bw() + theme(axis.text=element_text(colour="grey30",size=12),
									axis.title=element_text(colour="grey30",size=14),
									plot.title=element_text(size=14,hjust=0.5),
									plot.subtitle=element_text(size=12,hjust=0.5),
									legend.text=element_text(size=12,colour="grey30"),
									legend.title=element_text(size=12,colour="grey30")))


## load targets file for all cells. Columns required: 
targets <- read.delim(SHINYREPS_TARGET)
#group.vars <- colnames(targets)[!colnames(targets) %in% c("sample", "barcode", "row", "col")] 
#targets[,group.vars] <- lapply(targets[,group.vars], factor)

# print targets file
DT::datatable(targets, caption="targets file (sample sheet) used for analysis")

```

# Quality control

## FastQC of all reads 

The raw sequence reads of all samples are analysed with the popular FastQC tool (http://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

1. The "Read qualities" Box-Whisker plots show the range of quality values across all base positions:
    (i) The central red line is the median value,
    (ii) The yellow box represents the inter-quartile range (25-75%),
    (iii) The upper and lower whiskers represent the 10% and 90% points,
    (iv) The blue line represents the mean quality.
The y-axis on the graph shows the Phred quality scores, which are logarithmically related to the base-calling error probabilities. The higher the score the better the base call. The background of the graph divides the y axis into very good quality calls (green), calls of reasonable quality (orange), and calls of poor quality (red). Typically, the majority of calls on all base positions fall into the green area.

2. The "Sequence bias" plots show the proportion of each base (% G, A, T and C) at each position. In a random library there would be little difference between the positions of a sequence run, so the lines in this plot should run parallel with each other. But most RNA-seq libraries show sequence imbalance in the first 10-12 read positions due to RT priming biases, which should however look fairly similar in all samples.

3. The "GC Content" plots show the GC% distribution of all reads (red lines) and the ideal normal distribution based on the data (blue lines). Typically, the red and blue lines tightly overlap and look essentially the same in all samples. An unusually shaped distribution could indicate a contaminated library.

```{r FastQC_paragraph1, echo=F, results='asis', error=F, warning=F, message=F}
##### parameters to set:
# Select samples for which you would like to include fastqc results in the report. For single cell RNA-Seq with many cells, 
# you may want to restrict the total number of plots. If you provide a regular expression in 'samplePattern' only those 
# filenames will be included which match this expression, e.g. setting samplePattern="R1" yields only those fastq 
# files containing read1 of a read pair. This is recomended e.g. for MARS-Seq, where Read2 contains barcode information only.
# For samplePattern=NULL (default) all fastq files will be included.
# If you want to exclude samples according to the given certain pattern (instead of including), set exclude=T. 
# This is e.g. for excluding the trimmed fastq files which contain "cutadapt" in the file name (see below). 
# If you set argument 'maxno', the maximum sample number will be restricted accordingly to the first 'maxno' plots. 
#
# NOTE: this should be moved as parameters into shinyreports.vars.groovy
samplePattern="_R1\\."
excludePattern=F  # default FALSE
maxno=NULL        # default NULL
##### 

cat(DEhelper.Fastqc(web=F, samplePattern=samplePattern, exclude=excludePattern, maxno=maxno), sep="\n")
# These options also apply for the helper functions following below
```


## Read de-duplication and feature counting with UMI-tools

UMI-tools counts the mapped reads per feature obtained after de-duplication. UMI-tools uses the uniquely mapped reads from STAR as well as as the multimapped reads (but not those which are mapped to too many loci) as input reads.

```{r umicount, echo=F, results='asis', error=F, warning=F, message=F, eval=F}
##### parameters to set:
# define subset of umicount log files if desired (samplePattern=NULL (default) includes all files).
samplePattern=NULL
# define categories of targets.txt to be used for dot color in the plot (one plot per element of colorByFactor will be created). 
# The function will try to map the data from targets$sample to the umicount log file names (for this, the unique part of 
# targets$sample must be a substring of the file name). If you have umi-tools log files from pooled fastq files but cell-wise 
# information in targets.txt as in MARS-Seq, you may provide another customized targets object in 'targetsdf'. 
# Otherwise pruned file names will be used by default.
colorByFactor = NULL # NULL by default
targetsdf=targets
#####

DEhelper.umicount(samplePattern=samplePattern, colorByFactor=colorByFactor, targetsdf=targetsdf)
```






```{r store_data, echo=F, results='hide', error=F, warning=F, message=F, eval=F}

# store count data (all and by pool)
dir_countdata <- file.path("report_files", "count_data")

cat("count data matrices are stored in", dir_countdata, "\n")

countslist <- lapply(c("counts", "logcounts", "logcounts_not_adjusted"), function(x) {

  if (!file.exists(file.path(dir_countdata, x))) {dir.create(file.path(dir_countdata, x), recursive = T)}

  write.table(data.frame(gene=rownames(sce), symbol=rowData(sce)$SYMBOL, assay(sce, x)), 
              file =file.path(dir_countdata, x, paste0(x, "_all.txt")), sep="\t", quote = F, row.names = F)

  lapply(unique(colData(sce)$pool), function(y) {
    sce.subset <- sce[, colData(sce)$pool == y]
    write.table(data.frame(gene=rownames(sce.subset), symbol=rowData(sce.subset)$SYMBOL, assay(sce.subset, x)), 
                file = file.path(dir_countdata, x, paste0(x, "_", y, ".txt")), sep="\t", quote = F, row.names = F)
  })
})

# save workspace
save.image("report_files/WS.RData")

# create files for shiny app 
library(shinydashboard)

dir_shinyapp <- file.path("report_files", "shinyapp")
if (!file.exists(dir_shinyapp)) {dir.create(dir_shinyapp) }
  top.hvg <- hvg.cor.list[["all"]]$top.hvg
  genes <- gene.names[gene.names$gene_id %in% rownames(sce), ]
  save(sce, genes, top.hvg, file=file.path(dir_shinyapp, "data4shiny.RData"))
  
```




# Used tools and versions for this analysis ##

The following tools were used for data processing:

```{r ToolVersions_paragraph, echo=F, results='asis', error=F, warning=F, message=F}
cat(Toolhelper.ToolVersions(), sep="\n")
```

R session info:

```{r R_sessionInfo, echo=F, results='asis', error=F, warning=F, message=F}
sessionInfo()
```

</div>

